<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.262">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Evgenii O. Tretiakov">
<meta name="dcterms.date" content="2023-08-01">

<title>CellChat signalling analysis of Arcuate Nucleus/Median Eminence astrocytes dataset from Lutomska LM et al 2022</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="12-cellchat_files/libs/clipboard/clipboard.min.js"></script>
<script src="12-cellchat_files/libs/quarto-html/quarto.js"></script>
<script src="12-cellchat_files/libs/quarto-html/popper.min.js"></script>
<script src="12-cellchat_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="12-cellchat_files/libs/quarto-html/anchor.min.js"></script>
<link href="12-cellchat_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="12-cellchat_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="12-cellchat_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="12-cellchat_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="12-cellchat_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

<link href="12-cellchat_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="12-cellchat_files/libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#load-data-and-setup-parameters" id="toc-load-data-and-setup-parameters" class="nav-link active" data-scroll-target="#load-data-and-setup-parameters">Load data and setup parameters</a>
  <ul class="collapse">
  <li><a href="#set-paths" id="toc-set-paths" class="nav-link" data-scroll-target="#set-paths">Set paths</a></li>
  <li><a href="#load-helper-functions-and-gene-sets" id="toc-load-helper-functions-and-gene-sets" class="nav-link" data-scroll-target="#load-helper-functions-and-gene-sets">Load helper functions and gene-sets</a></li>
  <li><a href="#set-fixed-variables" id="toc-set-fixed-variables" class="nav-link" data-scroll-target="#set-fixed-variables">Set fixed variables</a></li>
  </ul></li>
  <li><a href="#load-selected-astrocytes-data-from-lutomska-lm-et-al-2022" id="toc-load-selected-astrocytes-data-from-lutomska-lm-et-al-2022" class="nav-link" data-scroll-target="#load-selected-astrocytes-data-from-lutomska-lm-et-al-2022">Load selected astrocytes data from Lutomska LM et al (2022)</a>
  <ul class="collapse">
  <li><a href="#convert-adata-object-to-r-anndatar6-object." id="toc-convert-adata-object-to-r-anndatar6-object." class="nav-link" data-scroll-target="#convert-adata-object-to-r-anndatar6-object.">Convert adata object to R AnnDataR6 object.</a></li>
  </ul></li>
  <li><a href="#plot-cellchat-network-of-hfhsd-and-control-mice" id="toc-plot-cellchat-network-of-hfhsd-and-control-mice" class="nav-link" data-scroll-target="#plot-cellchat-network-of-hfhsd-and-control-mice">Plot CellChat network of HFHSD and control mice</a>
  <ul class="collapse">
  <li><a href="#control" id="toc-control" class="nav-link" data-scroll-target="#control">Control</a></li>
  <li><a href="#hfhsd" id="toc-hfhsd" class="nav-link" data-scroll-target="#hfhsd">HFHSD</a></li>
  </ul></li>
  <li><a href="#compare-cellchat-network-of-hfhsd-and-control-mice" id="toc-compare-cellchat-network-of-hfhsd-and-control-mice" class="nav-link" data-scroll-target="#compare-cellchat-network-of-hfhsd-and-control-mice">Compare CellChat network of HFHSD and control mice</a>
  <ul class="collapse">
  <li><a href="#compare-the-total-number-of-interactions-and-interaction-strength" id="toc-compare-the-total-number-of-interactions-and-interaction-strength" class="nav-link" data-scroll-target="#compare-the-total-number-of-interactions-and-interaction-strength">Compare the total number of interactions and interaction strength</a></li>
  <li><a href="#compare-the-number-of-interactions-and-interaction-strength-among-different-cell-populations" id="toc-compare-the-number-of-interactions-and-interaction-strength-among-different-cell-populations" class="nav-link" data-scroll-target="#compare-the-number-of-interactions-and-interaction-strength-among-different-cell-populations">Compare the number of interactions and interaction strength among different cell populations</a></li>
  </ul></li>
  <li><a href="#identify-the-conserved-and-context-specific-signaling-pathways" id="toc-identify-the-conserved-and-context-specific-signaling-pathways" class="nav-link" data-scroll-target="#identify-the-conserved-and-context-specific-signaling-pathways">Identify the conserved and context-specific signaling pathways</a>
  <ul class="collapse">
  <li><a href="#identify-signaling-networks-with-larger-or-less-difference-as-well-as-signaling-groups-based-on-their-functionalstructure-similarity" id="toc-identify-signaling-networks-with-larger-or-less-difference-as-well-as-signaling-groups-based-on-their-functionalstructure-similarity" class="nav-link" data-scroll-target="#identify-signaling-networks-with-larger-or-less-difference-as-well-as-signaling-groups-based-on-their-functionalstructure-similarity">Identify signaling networks with larger (or less) difference as well as signaling groups based on their functional/structure similarity</a></li>
  <li><a href="#compute-and-visualize-the-pathway-distance-in-the-learned-joint-manifold" id="toc-compute-and-visualize-the-pathway-distance-in-the-learned-joint-manifold" class="nav-link" data-scroll-target="#compute-and-visualize-the-pathway-distance-in-the-learned-joint-manifold">Compute and visualize the pathway distance in the learned joint manifold</a></li>
  <li><a href="#identify-and-visualize-the-conserved-and-context-specific-signaling-pathways" id="toc-identify-and-visualize-the-conserved-and-context-specific-signaling-pathways" class="nav-link" data-scroll-target="#identify-and-visualize-the-conserved-and-context-specific-signaling-pathways">Identify and visualize the conserved and context-specific signaling pathways</a></li>
  <li><a href="#identify-dysfunctional-signaling-by-using-differential-expression-analysis" id="toc-identify-dysfunctional-signaling-by-using-differential-expression-analysis" class="nav-link" data-scroll-target="#identify-dysfunctional-signaling-by-using-differential-expression-analysis">Identify dysfunctional signaling by using differential expression analysis</a></li>
  <li><a href="#visually-compare-cell-cell-communication-using-hierarchy-plot-circle-plot" id="toc-visually-compare-cell-cell-communication-using-hierarchy-plot-circle-plot" class="nav-link" data-scroll-target="#visually-compare-cell-cell-communication-using-hierarchy-plot-circle-plot">Visually compare cell-cell communication using Hierarchy plot, Circle plot</a></li>
  </ul></li>
  <li><a href="#compare-the-signaling-gene-expression-distribution-between-different-datasets" id="toc-compare-the-signaling-gene-expression-distribution-between-different-datasets" class="nav-link" data-scroll-target="#compare-the-signaling-gene-expression-distribution-between-different-datasets">Compare the signaling gene expression distribution between different datasets</a></li>
  <li><a href="#save-the-merged-cellchat-object" id="toc-save-the-merged-cellchat-object" class="nav-link" data-scroll-target="#save-the-merged-cellchat-object">Save the merged CellChat object</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">CellChat signalling analysis of Arcuate Nucleus/Median Eminence astrocytes dataset from Lutomska LM et al 2022</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Evgenii O. Tretiakov </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 1, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="load-data-and-setup-parameters" class="level2">
<h2 class="anchored" data-anchor-id="load-data-and-setup-parameters">Load data and setup parameters</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load tidyverse infrastructure packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(future)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(here)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidyverse)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(magrittr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(stringr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(skimr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(viridis)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages for scRNA-seq analysis and visualisation</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(cowplot)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(patchwork)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggstatsplot)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(sceasy)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(anndata)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Seurat)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(SeuratDisk)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(CellChat)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scCustomize)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scanpy"</span>, <span class="at">convert =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="set-paths" class="level3">
<h3 class="anchored" data-anchor-id="set-paths">Set paths</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>src_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"code"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"data"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"output"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>plots_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(output_dir, <span class="st">"figures/"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>tables_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(output_dir, <span class="st">"tables/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="load-helper-functions-and-gene-sets" class="level3">
<h3 class="anchored" data-anchor-id="load-helper-functions-and-gene-sets">Load helper functions and gene-sets</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(src_dir, <span class="st">"genes.R"</span>))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(src_dir, <span class="st">"functions.R"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="set-fixed-variables" class="level3">
<h3 class="anchored" data-anchor-id="set-fixed-variables">Set fixed variables</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>reseed <span class="ot">&lt;-</span> <span class="dv">42</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="at">seed =</span> reseed)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters for parallel execution</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>n_cores <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="st">"multisession"</span>, <span class="at">workers =</span> n_cores)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">future.globals.maxSize =</span> <span class="dv">200000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">future.rng.onMisuse =</span> <span class="st">"ignore"</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>multisession:
- args: function (..., workers = 8, envir = parent.frame())
- tweaked: TRUE
- call: plan("multisession", workers = n_cores)</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot2 theme</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(ggmin<span class="sc">::</span><span class="fu">theme_powerpoint</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>bioproject <span class="ot">&lt;-</span> <span class="st">"PRJNA847050"</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>project <span class="ot">&lt;-</span> <span class="st">"lutomska2022-Arc"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>cb_fpr <span class="ot">&lt;-</span> <span class="fl">0.001</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>low_cutoff_gene <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>high_cutoff_gene <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>high_cutoff_gene <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>low_cutoff_umis <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>low_cutoff_umis <span class="ot">&lt;-</span> <span class="sc">-</span><span class="cn">Inf</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>high_cutoff_umis <span class="ot">&lt;-</span> <span class="dv">45000</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>high_cutoff_pc_mt <span class="ot">&lt;-</span> <span class="dv">25</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>high_cutoff_pc_ribo <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>high_cutoff_pc_hb <span class="ot">&lt;-</span> <span class="fl">0.1</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>high_cutoff_doublet_score <span class="ot">&lt;-</span> <span class="fl">0.33</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>high_cutoff_complexity <span class="ot">&lt;-</span> <span class="fl">0.8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="load-selected-astrocytes-data-from-lutomska-lm-et-al-2022" class="level2">
<h2 class="anchored" data-anchor-id="load-selected-astrocytes-data-from-lutomska-lm-et-al-2022">Load selected astrocytes data from Lutomska LM et al (2022)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>anndata <span class="ot">&lt;-</span> sc<span class="sc">$</span><span class="fu">read</span>(<span class="fu">here</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  data_dir,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sprintf</span>(<span class="st">"class_cello/%s-whole_dataset-%s-cello_annotation.h5ad"</span>, bioproject, cb_fpr)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>astrodata <span class="ot">&lt;-</span> sc<span class="sc">$</span><span class="fu">read</span>(<span class="fu">here</span>(</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  data_dir,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sprintf</span>(<span class="st">"class_cello/%s-astrocytes_dataset-%s-initial_selection.h5ad"</span>, bioproject, cb_fpr)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>astrocytes_idx <span class="ot">&lt;-</span> <span class="fu">py_to_r</span>(astrodata<span class="sc">$</span>obs_names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="convert-adata-object-to-r-anndatar6-object." class="level3">
<h3 class="anchored" data-anchor-id="convert-adata-object-to-r-anndatar6-object.">Convert adata object to R AnnDataR6 object.</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>adata <span class="ot">&lt;-</span> <span class="fu">py_to_r</span>(anndata)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AnnDataR6" "R6"       </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(adata<span class="sc">$</span>X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "dgCMatrix"
attr(,"package")
[1] "Matrix"</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>AnnData object with n_obs × n_vars = 14594 × 24746
    obs: 'nCount_RAW', 'nFeature_RAW', 'nCount_RNA', 'nFeature_RNA', 'orig.ident', 'nFeature_Diff', 'nCount_Diff', 'percent_mito', 'percent_ribo', 'percent_mito_ribo', 'percent_hb', 'log10GenesPerUMI', 'cell_name', 'barcode', 'latent_RT_efficiency', 'latent_cell_probability', 'latent_scale', 'doublet_score', 'predicted_doublets', 'QC', 'var_regex', 'RNA_snn_res.0.5', 'RNA_snn_res.0.71728703859461', 'RNA_snn_res.1.18939286532015', 'RNA_snn_res.3.00000100000001', 'seurat_clusters', 'k_tree', 'comb_clstr1', 'S.Score', 'G2M.Score', 'Phase', 'nCount_SCT', 'nFeature_SCT', 'SCT_snn_res.1', 'SCT_snn_res.1.10031502002514', 'SCT_snn_res.1.2198400795032', 'SCT_snn_res.1.36467746508232', 'SCT_snn_res.1.54382016547197', 'SCT_snn_res.1.77109349840298', 'SCT_snn_res.2.0689045782524', 'SCT_snn_res.2.47612417203789', 'SCT_snn_res.3.06661007679595', 'SCT_snn_res.4.00000100000002', 'bioproject', 'project', 'model', 'tech', 'region', 'sex', 'stage', 'libname', 'expbtch', 'condit', 'ora_celltype'
    var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'vst.variable'
    uns: 'k_tree_colors', 'name', 'ora_celltype_colors'
    obsm: 'X_pacmap', 'X_pca', 'X_umap', 'ora_estimate', 'ora_pvals'</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>srt_path <span class="ot">&lt;-</span> <span class="fu">here</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  data_dir,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sprintf</span>(<span class="st">"%s-whole_dataset-%s-cello_annotation.h5ad"</span>, bioproject, cb_fpr)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>expr_mtx <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">as.matrix</span>(adata<span class="sc">$</span>raw<span class="sc">$</span>X))</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expr_mtx) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(adata<span class="sc">$</span>X)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_mtx) <span class="ot">&lt;-</span> adata<span class="sc">$</span>var_names</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>srt <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  expr_mtx,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"RNA"</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"individual_hypothalamic_nuclei_astrocytes_evaluation_dataset"</span>,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">meta.data =</span> <span class="fu">as.data.frame</span>(adata<span class="sc">$</span>obs),</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="fu">colnames</span>(adata<span class="sc">$</span>X)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt) <span class="ot">&lt;-</span> <span class="st">"ora_celltype"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>srt <span class="ot">&lt;-</span> <span class="fu">subset</span>(srt, <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Astrocytes"</span>, <span class="st">"Neurons"</span>))</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt) <span class="ot">&lt;-</span> <span class="st">"expbtch"</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>srt_keep <span class="ot">&lt;-</span> srt</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>srt <span class="ot">&lt;-</span> <span class="fu">subset</span>(srt, <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"hfd_05"</span>))</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>srt<span class="sc">$</span>libname <span class="ot">&lt;-</span> <span class="fu">factor</span>(srt<span class="sc">$</span>expbtch, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"control"</span> <span class="ot">=</span> <span class="st">"control"</span>, <span class="st">"HFHSD"</span> <span class="ot">=</span> <span class="st">"hfd_05"</span>))</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt) <span class="ot">&lt;-</span> <span class="st">"libname"</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt) <span class="ot">&lt;-</span> <span class="st">"k_tree"</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(srt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24746 features across 4057 samples within 1 assay 
Active assay: RNA (24746 features, 0 variable features)</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(adata, astrodata, anndata, expr_mtx)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">gc</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt, <span class="at">cells =</span> <span class="fu">WhichCells</span>(srt, <span class="at">expression =</span> (Pomc <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">&amp;</span> ora_celltype <span class="sc">==</span> <span class="st">"Neurons"</span>))) <span class="ot">&lt;-</span> <span class="st">"POMC neurons"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt, <span class="at">cells =</span> <span class="fu">WhichCells</span>(srt, <span class="at">expression =</span> ((Agrp <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">|</span> Npy <span class="sc">&gt;</span> <span class="dv">10</span>) <span class="sc">&amp;</span> ora_celltype <span class="sc">==</span> <span class="st">"Neurons"</span>))) <span class="ot">&lt;-</span> <span class="st">"AGRP/NPY neurons"</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt, <span class="at">cells =</span> <span class="fu">WhichCells</span>(srt, <span class="at">expression =</span> (Insr <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> cell_name <span class="sc">%in%</span> astrocytes_idx))) <span class="ot">&lt;-</span> <span class="st">"Astro_INSR+"</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(srt, <span class="at">cells =</span> <span class="fu">WhichCells</span>(srt, <span class="at">expression =</span> (Insr <span class="sc">==</span> <span class="dv">0</span> <span class="sc">&amp;</span> cell_name <span class="sc">%in%</span> astrocytes_idx))) <span class="ot">&lt;-</span> <span class="st">"Astro_INSR-"</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>srt <span class="ot">&lt;-</span> <span class="fu">subset</span>(srt, <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Astro_INSR+"</span>, <span class="st">"Astro_INSR-"</span>, <span class="st">"POMC neurons"</span>, <span class="st">"AGRP/NPY neurons"</span>))</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>srt<span class="sc">$</span>goi <span class="ot">&lt;-</span> <span class="fu">Idents</span>(srt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>srt.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(srt, <span class="at">split.by =</span> <span class="st">"libname"</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>srt.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(NormalizeData)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>srt.list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>$hfd_05
An object of class Seurat 
24746 features across 1111 samples within 1 assay 
Active assay: RNA (24746 features, 0 variable features)

$control
An object of class Seurat 
24746 features across 1201 samples within 1 assay 
Active assay: RNA (24746 features, 0 variable features)</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>SrtCellChat <span class="ot">&lt;-</span> <span class="cf">function</span>(seurat_object) {</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  data.input <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(seurat_object, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">slot =</span> <span class="st">"data"</span>) <span class="co"># normalized data matrix</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  labels <span class="ot">&lt;-</span> <span class="fu">Idents</span>(seurat_object)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  meta <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">labels =</span> labels, <span class="at">row.names =</span> <span class="fu">names</span>(labels)) <span class="co"># create a dataframe of the cell labels</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  cellchat <span class="ot">&lt;-</span> <span class="fu">createCellChat</span>(<span class="at">object =</span> data.input, <span class="at">meta =</span> meta, <span class="at">group.by =</span> <span class="st">"labels"</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cellchat)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>object.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> srt.list, <span class="at">FUN =</span> SrtCellChat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Create a CellChat object from a data matrix"
Set cell identities for the new CellChat object 
The cell groups used for CellChat analysis are  Astro_INSR- Astro_INSR+ AGRP/NPY neurons POMC neurons 
[1] "Create a CellChat object from a data matrix"
Set cell identities for the new CellChat object 
The cell groups used for CellChat analysis are  Astro_INSR- Astro_INSR+ AGRP/NPY neurons POMC neurons </code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>CellChatDB <span class="ot">&lt;-</span> CellChatDB.mouse <span class="co"># running on mouse data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">showDatabaseCategory</span>(CellChatDB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchatdb-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(CellChatDB<span class="sc">$</span>interaction)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 2,019
Columns: 11
$ interaction_name   &lt;chr&gt; "TGFB1_TGFBR1_TGFBR2", "TGFB2_TGFBR1_TGFBR2", "TGFB…
$ pathway_name       &lt;chr&gt; "TGFb", "TGFb", "TGFb", "TGFb", "TGFb", "TGFb", "TG…
$ ligand             &lt;chr&gt; "Tgfb1", "Tgfb2", "Tgfb3", "Tgfb1", "Tgfb1", "Tgfb2…
$ receptor           &lt;chr&gt; "TGFbR1_R2", "TGFbR1_R2", "TGFbR1_R2", "ACVR1B_TGFb…
$ agonist            &lt;chr&gt; "TGFb agonist", "TGFb agonist", "TGFb agonist", "TG…
$ antagonist         &lt;chr&gt; "TGFb antagonist", "TGFb antagonist", "TGFb antagon…
$ co_A_receptor      &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "", "",…
$ co_I_receptor      &lt;chr&gt; "TGFb inhibition receptor", "TGFb inhibition recept…
$ evidence           &lt;chr&gt; "KEGG: mmu04350", "KEGG: mmu04350", "KEGG: mmu04350…
$ annotation         &lt;chr&gt; "Secreted Signaling", "Secreted Signaling", "Secret…
$ interaction_name_2 &lt;chr&gt; "Tgfb1 - (Tgfbr1+Tgfbr2)", "Tgfb2  - (Tgfbr1+Tgfbr2…</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>skimr<span class="sc">::</span><span class="fu">skim</span>(CellChatDB<span class="sc">$</span>interaction)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">CellChatDB$interaction</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">2019</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">character</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 24%">
<col style="width: 12%">
<col style="width: 18%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 7%">
<col style="width: 11%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">min</th>
<th style="text-align: right;">max</th>
<th style="text-align: right;">empty</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: right;">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">interaction_name</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2019</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">pathway_name</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">229</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ligand</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">577</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">receptor</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">510</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">agonist</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">1676</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">antagonist</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">1563</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">co_A_receptor</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">1637</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">co_I_receptor</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">1530</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">evidence</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">169</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">annotation</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">interaction_name_2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">26</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2019</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>janitor<span class="sc">::</span><span class="fu">tabyl</span>(CellChatDB<span class="sc">$</span>interaction, annotation)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["annotation"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]},{"label":["percent"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"Cell-Cell Contact","2":"378","3":"0.1872214"},{"1":"ECM-Receptor","2":"432","3":"0.2139673"},{"1":"Secreted Signaling","2":"1209","3":"0.5988113"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>CellChatDB.use <span class="ot">&lt;-</span> CellChatDB <span class="co"># simply use the default CellChatDB</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">gc</span>())</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="st">"sequential"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">gc</span>())</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="st">"multisession"</span>, <span class="at">workers =</span> n_cores)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(<span class="fu">here</span>(data_dir, <span class="st">"cellchat-lutomska2022-HFHSD5_and_control.rds"</span>))) {</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    object.list <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(<span class="at">file =</span> <span class="fu">here</span>(data_dir, <span class="st">"cellchat-lutomska2022-HFHSD5_and_control.rds"</span>))</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># set the used database in the object</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>  object.list<span class="sc">$</span>hfd_05<span class="sc">@</span>DB <span class="ot">&lt;-</span> CellChatDB.use</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  object.list<span class="sc">$</span>control<span class="sc">@</span>DB <span class="ot">&lt;-</span> CellChatDB.use</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(subsetData)</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(identifyOverExpressedGenes)</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(identifyOverExpressedInteractions)</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(projectData, PPI.mouse)</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(computeCommunProb, <span class="at">raw.use =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(filterCommunication, <span class="at">min.cells =</span> <span class="dv">5</span>)</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(computeCommunProbPathway)</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(aggregateNet)</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>  object.list <span class="sc">%&lt;&gt;%</span> <span class="fu">map</span>(netAnalysis_computeCentrality)</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_rds</span>(object.list, <span class="at">file =</span> <span class="fu">here</span>(data_dir, <span class="st">"cellchat-lutomska2022-HFHSD5_and_control.rds"</span>))</span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="plot-cellchat-network-of-hfhsd-and-control-mice" class="level2">
<h2 class="anchored" data-anchor-id="plot-cellchat-network-of-hfhsd-and-control-mice">Plot CellChat network of HFHSD and control mice</h2>
<section id="control" class="level3">
<h3 class="anchored" data-anchor-id="control">Control</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> object.list<span class="sc">$</span>control</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>groupSize <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">table</span>(cellchat<span class="sc">@</span>idents))</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>count, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T, <span class="at">title.name =</span> <span class="st">"Number of interactions"</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>weight, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T, <span class="at">title.name =</span> <span class="st">"Interaction weights/strength"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-control-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> cellchat<span class="sc">@</span>net<span class="sc">$</span>weight</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mat)) {</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  mat2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(mat), <span class="at">ncol =</span> <span class="fu">ncol</span>(mat), <span class="at">dimnames =</span> <span class="fu">dimnames</span>(mat))</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  mat2[i, ] <span class="ot">&lt;-</span> mat[i, ]</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">netVisual_circle</span>(mat2, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">edge.weight.max =</span> <span class="fu">max</span>(mat), <span class="at">title.name =</span> <span class="fu">rownames</span>(mat)[i])</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-per-group-control-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="hfhsd" class="level3">
<h3 class="anchored" data-anchor-id="hfhsd">HFHSD</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> object.list<span class="sc">$</span>hfd_05</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>groupSize <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">table</span>(cellchat<span class="sc">@</span>idents))</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>count, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T, <span class="at">title.name =</span> <span class="st">"Number of interactions"</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>weight, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T, <span class="at">title.name =</span> <span class="st">"Interaction weights/strength"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-HFHSD-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> cellchat<span class="sc">@</span>net<span class="sc">$</span>weight</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mat)) {</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  mat2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(mat), <span class="at">ncol =</span> <span class="fu">ncol</span>(mat), <span class="at">dimnames =</span> <span class="fu">dimnames</span>(mat))</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  mat2[i, ] <span class="ot">&lt;-</span> mat[i, ]</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">netVisual_circle</span>(mat2, <span class="at">vertex.weight =</span> groupSize, <span class="at">weight.scale =</span> T, <span class="at">edge.weight.max =</span> <span class="fu">max</span>(mat), <span class="at">title.name =</span> <span class="fu">rownames</span>(mat)[i])</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-per-group-HFHSD-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="compare-cellchat-network-of-hfhsd-and-control-mice" class="level2">
<h2 class="anchored" data-anchor-id="compare-cellchat-network-of-hfhsd-and-control-mice">Compare CellChat network of HFHSD and control mice</h2>
<section id="compare-the-total-number-of-interactions-and-interaction-strength" class="level3">
<h3 class="anchored" data-anchor-id="compare-the-total-number-of-interactions-and-interaction-strength">Compare the total number of interactions and interaction strength</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>object.list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">control =</span> object.list<span class="sc">$</span>control, <span class="at">HFHSD =</span> object.list<span class="sc">$</span>hfd_05)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">mergeCellChat</span>(object.list, <span class="at">add.names =</span> <span class="fu">names</span>(object.list))</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">compareInteractions</span>(cellchat, <span class="at">show.legend =</span> F, <span class="at">group =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">compareInteractions</span>(cellchat, <span class="at">show.legend =</span> F, <span class="at">group =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">measure =</span> <span class="st">"weight"</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> gg2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-1.png" class="img-fluid figure-img" width="1200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="compare-the-number-of-interactions-and-interaction-strength-among-different-cell-populations" class="level3">
<h3 class="anchored" data-anchor-id="compare-the-number-of-interactions-and-interaction-strength-among-different-cell-populations">Compare the number of interactions and interaction strength among different cell populations</h3>
<section id="differential-number-of-interactions-or-interaction-strength-among-different-cell-populations" class="level4">
<h4 class="anchored" data-anchor-id="differential-number-of-interactions-or-interaction-strength-among-different-cell-populations">Differential number of interactions or interaction strength among different cell populations</h4>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_diffInteraction</span>(cellchat, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_diffInteraction</span>(cellchat, <span class="at">weight.scale =</span> T, <span class="at">measure =</span> <span class="st">"weight"</span>, <span class="at">label.edge =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-strength-per-group-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">netVisual_heatmap</span>(cellchat)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">netVisual_heatmap</span>(cellchat, <span class="at">measure =</span> <span class="st">"weight"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> gg2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-strength-per-group-heatmap-1.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>weight.max <span class="ot">&lt;-</span> <span class="fu">getMaxWeight</span>(object.list, <span class="at">attribute =</span> <span class="fu">c</span>(<span class="st">"idents"</span>, <span class="st">"count"</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(object.list)) {</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">netVisual_circle</span>(object.list[[i]]<span class="sc">@</span>net<span class="sc">$</span>count, <span class="at">weight.scale =</span> T, <span class="at">label.edge =</span> T, <span class="at">edge.weight.max =</span> weight.max[<span class="dv">2</span>], <span class="at">edge.width.max =</span> <span class="dv">12</span>, <span class="at">title.name =</span> <span class="fu">paste0</span>(<span class="st">"Number of interactions - "</span>, <span class="fu">names</span>(object.list)[i]))</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-strength-per-group-split-weight-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="compare-the-major-sources-and-targets-in-2d-space" class="level4">
<h4 class="anchored" data-anchor-id="compare-the-major-sources-and-targets-in-2d-space">Compare the major sources and targets in 2D space</h4>
<div class="cell" data-layout-align="center" data-fig.asp="0.618">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>num.link <span class="ot">&lt;-</span> <span class="fu">sapply</span>(object.list, <span class="cf">function</span>(x) {</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowSums</span>(x<span class="sc">@</span>net<span class="sc">$</span>count) <span class="sc">+</span> <span class="fu">colSums</span>(x<span class="sc">@</span>net<span class="sc">$</span>count) <span class="sc">-</span> <span class="fu">diag</span>(x<span class="sc">@</span>net<span class="sc">$</span>count)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>weight.MinMax <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(num.link), <span class="fu">max</span>(num.link)) <span class="co"># control the dot size in the different datasets</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(object.list)) {</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  gg[[i]] <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_scatter</span>(object.list[[i]], <span class="at">title =</span> <span class="fu">names</span>(object.list)[i], <span class="at">weight.MinMax =</span> weight.MinMax)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>patchwork<span class="sc">::</span><span class="fu">wrap_plots</span>(<span class="at">plots =</span> gg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-strength-incoming-outgoing-1.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-fig.asp="0.618">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingChanges_scatter</span>(cellchat, <span class="at">idents.use =</span> <span class="st">"Astro_INSR+"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingChanges_scatter</span>(cellchat, <span class="at">idents.use =</span> <span class="st">"Astro_INSR-"</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>gg3 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingChanges_scatter</span>(cellchat, <span class="at">idents.use =</span> <span class="st">"POMC neurons"</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>gg4 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingChanges_scatter</span>(cellchat, <span class="at">idents.use =</span> <span class="st">"AGRP/NPY neurons"</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>patchwork<span class="sc">::</span><span class="fu">wrap_plots</span>(<span class="at">plots =</span> <span class="fu">list</span>(gg1, gg2, gg3, gg4))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-compare-strength-incoming-outgoing-per-group-1.png" class="img-fluid figure-img" width="3900"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="identify-the-conserved-and-context-specific-signaling-pathways" class="level2">
<h2 class="anchored" data-anchor-id="identify-the-conserved-and-context-specific-signaling-pathways">Identify the conserved and context-specific signaling pathways</h2>
<section id="identify-signaling-networks-with-larger-or-less-difference-as-well-as-signaling-groups-based-on-their-functionalstructure-similarity" class="level3">
<h3 class="anchored" data-anchor-id="identify-signaling-networks-with-larger-or-less-difference-as-well-as-signaling-groups-based-on-their-functionalstructure-similarity">Identify signaling networks with larger (or less) difference as well as signaling groups based on their functional/structure similarity</h3>
<p>Functional similarity: High degree of functional similarity indicates major senders and receivers are similar, and it can be interpreted as the two signaling pathways or two ligand-receptor pairs exhibit similar and/or redundant roles. NB: Functional similarity analysis is not applicable to multiple datsets with different cell type composition.</p>
<p>Structural similarity: A structural similarity was used to compare their signaling network structure, without considering the similarity of senders and receivers. NB: Structural similarity analysis is applicable to multiple datsets with the same cell type composition or the vastly different cell type composition.</p>
<section id="identify-signaling-groups-based-on-their-functional-similarity" class="level4">
<h4 class="anchored" data-anchor-id="identify-signaling-groups-based-on-their-functional-similarity">Identify signaling groups based on their functional similarity</h4>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">computeNetSimilarityPairwise</span>(cellchat, <span class="at">type =</span> <span class="st">"functional"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Compute signaling network similarity for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">netEmbedding</span>(cellchat, <span class="at">type =</span> <span class="st">"functional"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Manifold learning of the signaling networks for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">netClustering</span>(cellchat, <span class="at">type =</span> <span class="st">"functional"</span>, <span class="at">do.parallel =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Classification learning of the signaling networks for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_embeddingPairwise</span>(cellchat, <span class="at">type =</span> <span class="st">"functional"</span>, <span class="at">label.size =</span> <span class="fl">3.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2D visualization of signaling networks from datasets 1 2 </code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-functional-similarity-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="identify-signaling-groups-based-on-structure-similarity" class="level4">
<h4 class="anchored" data-anchor-id="identify-signaling-groups-based-on-structure-similarity">Identify signaling groups based on structure similarity</h4>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">computeNetSimilarityPairwise</span>(cellchat, <span class="at">type =</span> <span class="st">"structural"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Compute signaling network similarity for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">netEmbedding</span>(cellchat, <span class="at">type =</span> <span class="st">"structural"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Manifold learning of the signaling networks for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">netClustering</span>(cellchat, <span class="at">type =</span> <span class="st">"structural"</span>, <span class="at">do.parallel =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Classification learning of the signaling networks for datasets 1 2 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_embeddingPairwise</span>(cellchat, <span class="at">type =</span> <span class="st">"structural"</span>, <span class="at">label.size =</span> <span class="fl">3.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2D visualization of signaling networks from datasets 1 2 </code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-structural-similarity-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="compute-and-visualize-the-pathway-distance-in-the-learned-joint-manifold" class="level3">
<h3 class="anchored" data-anchor-id="compute-and-visualize-the-pathway-distance-in-the-learned-joint-manifold">Compute and visualize the pathway distance in the learned joint manifold</h3>
<p>We can identify the signaling networks with larger (or less) difference based on their Euclidean distance in the shared two-dimensions space. Larger distance implies larger difference of the communication networks between two datasets in terms of either functional or structure similarity. NB: We only compute the distance of overlapped signaling pathways between two datasets. Those signaling pathways that are only identified in one dataset are not considered here. If there are more than three datasets, one can do pairwise comparisons by defining comparison in the function rankSimilarity.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rankSimilarity</span>(cellchat, <span class="at">type =</span> <span class="st">"functional"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Compute the distance of signaling networks between datasets 1 2 </code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-distance-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="identify-and-visualize-the-conserved-and-context-specific-signaling-pathways" class="level3">
<h3 class="anchored" data-anchor-id="identify-and-visualize-the-conserved-and-context-specific-signaling-pathways">Identify and visualize the conserved and context-specific signaling pathways</h3>
<p>By comparing the information flow/interaction strengh of each signaling pathway, we can identify signaling pathways, (i) turn off, (ii) decrease, (iii) turn on or (iv) increase, by change their information flow at one condition as compared to another condition.</p>
<section id="compare-the-overall-information-flow-of-each-signaling-pathway" class="level4">
<h4 class="anchored" data-anchor-id="compare-the-overall-information-flow-of-each-signaling-pathway">Compare the overall information flow of each signaling pathway</h4>
<p>We can identify the conserved and context-specific signaling pathways by simply comparing the information flow for each signaling pathway, which is defined by the sum of communication probability among all pairs of cell groups in the inferred network (i.e., the total weights in the network).</p>
<p>This bar graph can be plotted in a stacked mode or not. Significant signaling pathways were ranked based on differences in the overall information flow within the inferred networks between HFHSD and control mice arcuate nucleus of hypothalamus. The top signaling pathways colored red are enriched in HFHSD mice, and these colored blue were enriched in the control mice.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">rankNet</span>(cellchat, <span class="at">mode =</span> <span class="st">"comparison"</span>, <span class="at">stacked =</span> T, <span class="at">do.stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">rankNet</span>(cellchat, <span class="at">mode =</span> <span class="st">"comparison"</span>, <span class="at">stacked =</span> F, <span class="at">do.stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> gg2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-overall-flow-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="compare-outgoing-or-incoming-signaling-associated-with-each-cell-population" class="level4">
<h4 class="anchored" data-anchor-id="compare-outgoing-or-incoming-signaling-associated-with-each-cell-population">Compare outgoing (or incoming) signaling associated with each cell population</h4>
<p>The above analysis summarize the information from the outgoing and incoming signaling together. We can also compare the outgoing (or incoming) signaling pattern between two datasets, allowing to identify signaling pathways/ligand-receptors that exhibit different signaling patterns.</p>
<p>We can combine all the identified signaling pathways from different datasets and thus compare them side by side, including outgoing signaling, incoming signaling and overall signaling by aggregating outgoing and incoming signaling together. NB: rankNet also shows the comparison of overall signaling, but it does not show the signaling strength in specific cell populations.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co"># combining all the identified signaling pathways from different datasets</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>pathway.union <span class="ot">&lt;-</span> <span class="fu">union</span>(object.list[[i]]<span class="sc">@</span>netP<span class="sc">$</span>pathways, object.list[[i <span class="sc">+</span> <span class="dv">1</span>]]<span class="sc">@</span>netP<span class="sc">$</span>pathways)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>ht1 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i]], <span class="at">pattern =</span> <span class="st">"outgoing"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>ht2 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i <span class="sc">+</span> <span class="dv">1</span>]], <span class="at">pattern =</span> <span class="st">"outgoing"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i <span class="sc">+</span> <span class="dv">1</span>], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>)</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(ht1 <span class="sc">+</span> ht2, <span class="at">ht_gap =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-outgoing-flow-per-group-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>ht1 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i]], <span class="at">pattern =</span> <span class="st">"incoming"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>, <span class="at">color.heatmap =</span> <span class="st">"GnBu"</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>ht2 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i <span class="sc">+</span> <span class="dv">1</span>]], <span class="at">pattern =</span> <span class="st">"incoming"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i <span class="sc">+</span> <span class="dv">1</span>], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>, <span class="at">color.heatmap =</span> <span class="st">"GnBu"</span>)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(ht1 <span class="sc">+</span> ht2, <span class="at">ht_gap =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-incoming-flow-per-group-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>ht1 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i]], <span class="at">pattern =</span> <span class="st">"all"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>, <span class="at">color.heatmap =</span> <span class="st">"OrRd"</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>ht2 <span class="ot">&lt;-</span> <span class="fu">netAnalysis_signalingRole_heatmap</span>(object.list[[i <span class="sc">+</span> <span class="dv">1</span>]], <span class="at">pattern =</span> <span class="st">"all"</span>, <span class="at">signaling =</span> pathway.union, <span class="at">title =</span> <span class="fu">names</span>(object.list)[i <span class="sc">+</span> <span class="dv">1</span>], <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">14</span>, <span class="at">color.heatmap =</span> <span class="st">"OrRd"</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(ht1 <span class="sc">+</span> ht2, <span class="at">ht_gap =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-overall-flow-per-group-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="identify-dysfunctional-signaling-by-using-differential-expression-analysis" class="level3">
<h3 class="anchored" data-anchor-id="identify-dysfunctional-signaling-by-using-differential-expression-analysis">Identify dysfunctional signaling by using differential expression analysis</h3>
<p>The above method for identifying the upgulated and down-regulated signaling is perfomed by comparing the communication probability between two datasets for each L-R pair and each pair of cell groups. Alternative, we can identify the upgulated and down-regulated signaling ligand-receptor pairs based on the differential gene expression analysis. Specifically, we perform differential expression analysis between two biological conditions (HFHSD and control mice) for each cell group, and then obtain the upgulated and down-regulated signaling based on the fold change of ligands in the sender cells and receptors in the receiver cells. Such analysis can be done as follows.</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define a positive dataset, i.e., the dataset with positive fold change against the other dataset</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>pos.dataset <span class="ot">&lt;-</span> <span class="st">"HFHSD"</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># define a char name used for storing the results of differential expression analysis</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>features.name <span class="ot">&lt;-</span> pos.dataset</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co"># perform differential expression analysis</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">identifyOverExpressedGenes</span>(cellchat, <span class="at">group.dataset =</span> <span class="st">"datasets"</span>, <span class="at">pos.dataset =</span> pos.dataset, <span class="at">features.name =</span> features.name, <span class="at">only.pos =</span> <span class="cn">FALSE</span>, <span class="at">thresh.pc =</span> <span class="fl">0.1</span>, <span class="at">thresh.fc =</span> <span class="fl">0.1</span>, <span class="at">thresh.p =</span> <span class="dv">1</span>)</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Use the joint cell labels from the merged CellChat object</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co"># map the results of differential expression analysis onto the inferred cell-cell communications to easily manage/subset the ligand-receptor pairs of interest</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>net <span class="ot">&lt;-</span> <span class="fu">netMappingDEG</span>(cellchat, <span class="at">features.name =</span> features.name)</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the ligand-receptor pairs with upregulated ligands in HFHSD</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>net.up <span class="ot">&lt;-</span> <span class="fu">subsetCommunication</span>(cellchat, <span class="at">net =</span> net, <span class="at">datasets =</span> <span class="st">"HFHSD"</span>, <span class="at">ligand.logFC =</span> <span class="fl">0.2</span>, <span class="at">receptor.logFC =</span> <span class="cn">NULL</span>)</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the ligand-receptor pairs with upregulated ligands and upregulated recetptors in control, i.e.,downregulated in HFHSD</span></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>net.down <span class="ot">&lt;-</span> <span class="fu">subsetCommunication</span>(cellchat, <span class="at">net =</span> net, <span class="at">datasets =</span> <span class="st">"control"</span>, <span class="at">ligand.logFC =</span> <span class="sc">-</span><span class="fl">0.1</span>, <span class="at">receptor.logFC =</span> <span class="sc">-</span><span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Since the signaling genes in the net.up and net.down might be complex with multi-subunits, we can do further deconvolution to obtain the individual signaling genes</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">extractGeneSubsetFromPair</span>(net.up, cellchat)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>gene.down <span class="ot">&lt;-</span> <span class="fu">extractGeneSubsetFromPair</span>(net.down, cellchat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>pairLR.use.up <span class="ot">&lt;-</span> net.up[, <span class="st">"interaction_name"</span>, drop <span class="ot">=</span> F]</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">netVisual_bubble</span>(cellchat, <span class="at">pairLR.use =</span> pairLR.use.up, <span class="at">sources.use =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">targets.use =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="at">comparison =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">angle.x =</span> <span class="dv">90</span>, <span class="at">remove.isolate =</span> T, <span class="at">title.name =</span> <span class="fu">paste0</span>(<span class="st">"Up-regulated signaling in "</span>, <span class="fu">names</span>(object.list)[<span class="dv">2</span>]))</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Comparing communications on a merged object</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>pairLR.use.down <span class="ot">&lt;-</span> net.down[, <span class="st">"interaction_name"</span>, drop <span class="ot">=</span> F]</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">netVisual_bubble</span>(cellchat, <span class="at">pairLR.use =</span> pairLR.use.down, <span class="at">sources.use =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">targets.use =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="at">comparison =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">angle.x =</span> <span class="dv">90</span>, <span class="at">remove.isolate =</span> T, <span class="at">title.name =</span> <span class="fu">paste0</span>(<span class="st">"Down-regulated signaling in "</span>, <span class="fu">names</span>(object.list)[<span class="dv">2</span>]))</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Comparing communications on a merged object</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> gg2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-upregulated-ligand-receptor-pairs-12-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">netVisual_bubble</span>(cellchat, <span class="at">pairLR.use =</span> pairLR.use.up, <span class="at">sources.use =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="at">targets.use =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">comparison =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">angle.x =</span> <span class="dv">90</span>, <span class="at">remove.isolate =</span> T, <span class="at">title.name =</span> <span class="fu">paste0</span>(<span class="st">"Up-regulated signaling in "</span>, <span class="fu">names</span>(object.list)[<span class="dv">2</span>]))</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Comparing communications on a merged object</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>pairLR.use.down <span class="ot">&lt;-</span> net.down[, <span class="st">"interaction_name"</span>, drop <span class="ot">=</span> F]</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>gg2 <span class="ot">&lt;-</span> <span class="fu">netVisual_bubble</span>(cellchat, <span class="at">pairLR.use =</span> pairLR.use.down, <span class="at">sources.use =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="at">targets.use =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">comparison =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">angle.x =</span> <span class="dv">90</span>, <span class="at">remove.isolate =</span> T, <span class="at">title.name =</span> <span class="fu">paste0</span>(<span class="st">"Down-regulated signaling in "</span>, <span class="fu">names</span>(object.list)[<span class="dv">2</span>]))</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Comparing communications on a merged object</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> gg2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-upregulated-ligand-receptor-pairs-34-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
<p>Visualize the enriched ligands, signaling,or ligand-receptor pairs in one condition compared to another condition using wordcloud</p>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">computeEnrichmentScore</span>(net.down, <span class="at">species =</span> <span class="st">"mouse"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-wordcloud-enriched-downregulated-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visually-compare-cell-cell-communication-using-hierarchy-plot-circle-plot" class="level3">
<h3 class="anchored" data-anchor-id="visually-compare-cell-cell-communication-using-hierarchy-plot-circle-plot">Visually compare cell-cell communication using Hierarchy plot, Circle plot</h3>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>pathways.show.all <span class="ot">&lt;-</span> <span class="fu">intersect</span>(object.list<span class="sc">$</span>HFHSD<span class="sc">@</span>netP<span class="sc">$</span>pathways, object.list<span class="sc">$</span>control<span class="sc">@</span>netP<span class="sc">$</span>pathways)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pathways.show.all)) {</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  pathways.show <span class="ot">&lt;-</span> pathways.show.all[i]</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  ht <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tryCatch</span>(</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(object.list)) {</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>            ht[[j]] <span class="ot">&lt;-</span> <span class="fu">netVisual_heatmap</span>(object.list[[j]], <span class="at">signaling =</span> pathways.show, <span class="at">color.heatmap =</span> <span class="st">"Reds"</span>, <span class="at">title.name =</span> <span class="fu">paste</span>(pathways.show, <span class="st">"signaling "</span>, <span class="fu">names</span>(object.list)[j]))</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>        ComplexHeatmap<span class="sc">::</span><span class="fu">draw</span>(ht[[<span class="dv">1</span>]] <span class="sc">+</span> ht[[<span class="dv">2</span>]], <span class="at">ht_gap =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>))</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If an error occurs, silently skip to the next iteration</span></span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-1.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-2.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-3.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-4.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-5.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-6.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-7.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-8.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-9.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-10.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-11.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-12.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-13.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-14.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-15.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-16.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-17.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-18.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-19.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-20.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-21.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-22.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-23.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-24.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-25.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-26.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-27.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-28.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-29.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-30.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-31.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-32.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-hierarchy-33.png" class="img-fluid figure-img" width="2700"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pathways.show.all)) {</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tryCatch</span>(</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>        pathways.show <span class="ot">&lt;-</span> pathways.show.all[i]</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>        weight.max <span class="ot">&lt;-</span> <span class="fu">getMaxWeight</span>(object.list, <span class="at">slot.name =</span> <span class="fu">c</span>(<span class="st">"netP"</span>), <span class="at">attribute =</span> pathways.show) <span class="co"># control the edge weights across different datasets</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(object.list)) {</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">netVisual_aggregate</span>(object.list[[j]], <span class="at">signaling =</span> pathways.show, <span class="at">label.edge =</span> T, <span class="at">layout =</span> <span class="st">"circle"</span>, <span class="at">edge.weight.max =</span> weight.max[<span class="dv">1</span>], <span class="at">edge.width.max =</span> <span class="dv">10</span>, <span class="at">signaling.name =</span> <span class="fu">paste</span>(pathways.show, <span class="fu">names</span>(object.list)[j]))</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If an error occurs, silently skip to the next iteration</span></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-1.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-2.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-3.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-4.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-5.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-6.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-7.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-8.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-9.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-10.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-11.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-12.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-13.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-14.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-15.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-16.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-17.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-18.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-19.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-20.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-21.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-22.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-23.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-24.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-25.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-26.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-27.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-28.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-29.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-30.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-31.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-32.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12-cellchat_files/figure-html/plot-cellchat-net-pathway-compare-circle-33.png" class="img-fluid figure-img" width="4200"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="compare-the-signaling-gene-expression-distribution-between-different-datasets" class="level2">
<h2 class="anchored" data-anchor-id="compare-the-signaling-gene-expression-distribution-between-different-datasets">Compare the signaling gene expression distribution between different datasets</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>cellchat<span class="sc">@</span>meta<span class="sc">$</span>datasets <span class="ot">&lt;-</span> <span class="fu">factor</span>(cellchat<span class="sc">@</span>meta<span class="sc">$</span>datasets, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"HFHSD"</span>)) <span class="co"># set factor level</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pathways.show.all)) {</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tryCatch</span>(</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>        pathways.show <span class="ot">&lt;-</span> pathways.show.all[i]</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plotGeneExpression</span>(cellchat, <span class="at">signaling =</span> pathways.show, <span class="at">split.by =</span> <span class="st">"datasets"</span>, <span class="at">colors.ggplot =</span> T)</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If an error occurs, silently skip to the next iteration</span></span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="save-the-merged-cellchat-object" class="level2">
<h2 class="anchored" data-anchor-id="save-the-merged-cellchat-object">Save the merged CellChat object</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">write_rds</span>(cellchat, <span class="at">file =</span> <span class="fu">here</span>(data_dir, <span class="st">"cellchat-comparison-lutomska2022-HFHSD5_vs_control.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="session-information" class="level2">
<h2 class="anchored" data-anchor-id="session-information">Session information</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>sI <span class="ot">&lt;-</span> sessioninfo<span class="sc">::</span><span class="fu">session_info</span>()</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>sI<span class="sc">$</span>loadedOnly <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(sI, <span class="at">locale =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       Ubuntu 22.04.2 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Etc/UTC
 date     2023-08-01
 pandoc   2.19.2 @ /home/etretiakov/micromamba/bin/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package          * version     date (UTC) lib source
 abind              1.4-5       2016-07-21 [1] RSPM (R 4.2.2)
 anndata          * 0.7.5.6     2023-03-17 [1] RSPM (R 4.3.0)
 assertthat         0.2.1       2019-03-21 [1] CRAN (R 4.2.2)
 backports          1.4.1       2021-12-13 [1] RSPM (R 4.2.2)
 base64enc          0.1-3       2015-07-28 [1] RSPM (R 4.2.2)
 bayestestR         0.13.1      2023-04-07 [1] RSPM (R 4.3.0)
 beeswarm           0.4.0       2021-06-01 [1] RSPM (R 4.2.2)
 Biobase          * 2.60.0      2023-04-25 [1] Bioconductor
 BiocGenerics     * 0.46.0      2023-04-25 [1] Bioconductor
 BiocManager        1.30.21.1   2023-07-18 [1] RSPM (R 4.3.1)
 BiocNeighbors      1.18.0      2023-04-25 [1] Bioconductor
 BiocParallel       1.34.2      2023-05-22 [1] Bioconductor
 bit                4.0.5       2022-11-15 [1] RSPM (R 4.2.2)
 bit64              4.0.5       2020-08-30 [1] RSPM (R 4.2.2)
 broom              1.0.5       2023-06-09 [1] RSPM (R 4.3.0)
 Cairo              1.6-0       2022-07-05 [1] RSPM (R 4.2.2)
 car                3.1-2       2023-03-30 [1] RSPM (R 4.3.0)
 carData            3.0-5       2022-01-06 [1] RSPM (R 4.2.2)
 CellChat         * 1.6.1       2023-07-30 [1] Github (sqjin/CellChat@7ee94c3)
 circlize           0.4.15      2022-05-10 [1] RSPM (R 4.2.2)
 cli                3.6.1       2023-03-23 [1] RSPM (R 4.3.0)
 clue               0.3-64      2023-01-31 [1] RSPM (R 4.3.0)
 cluster            2.1.4       2022-08-22 [3] CRAN (R 4.2.1)
 coda               0.19-4      2020-09-30 [1] RSPM (R 4.2.2)
 codetools          0.2-19      2023-02-01 [3] CRAN (R 4.2.2)
 colorspace         2.1-0       2023-01-23 [1] RSPM (R 4.3.0)
 ComplexHeatmap   * 2.16.0      2023-04-25 [1] Bioconductor
 correlation        0.8.4       2023-04-28 [1] Github (easystats/correlation@3d5cd1e)
 cowplot          * 1.1.1       2020-12-30 [1] RSPM (R 4.2.2)
 crayon             1.5.2       2022-09-29 [1] RSPM (R 4.2.2)
 data.table         1.14.8      2023-02-17 [1] RSPM (R 4.3.0)
 datawizard         0.8.0       2023-06-16 [1] RSPM (R 4.3.0)
 DBI                1.1.3       2022-06-18 [1] RSPM (R 4.2.2)
 deldir             1.0-9       2023-05-17 [1] RSPM (R 4.3.0)
 digest             0.6.33      2023-07-07 [1] RSPM (R 4.3.1)
 doParallel         1.0.17      2022-02-07 [1] RSPM (R 4.2.2)
 dplyr            * 1.1.2       2023-04-20 [1] RSPM (R 4.3.0)
 ellipsis           0.3.2       2021-04-29 [1] RSPM (R 4.2.2)
 emmeans            1.8.6       2023-05-11 [1] RSPM (R 4.3.0)
 estimability       1.4.1       2022-08-05 [1] RSPM (R 4.3.0)
 evaluate           0.21        2023-05-05 [1] RSPM (R 4.3.0)
 fansi              1.0.4       2023-01-22 [1] RSPM (R 4.3.0)
 farver             2.1.1       2022-07-06 [1] RSPM (R 4.2.2)
 fastmap            1.1.1       2023-02-24 [1] RSPM (R 4.3.0)
 fitdistrplus       1.1-11      2023-04-25 [1] RSPM (R 4.3.0)
 FNN                1.1.3.2     2023-03-20 [1] RSPM (R 4.3.0)
 forcats          * 1.0.0       2023-01-29 [1] RSPM (R 4.3.0)
 foreach            1.5.2       2022-02-02 [1] RSPM (R 4.2.2)
 future           * 1.33.0      2023-07-01 [1] RSPM (R 4.3.1)
 future.apply       1.11.0      2023-05-21 [1] RSPM (R 4.3.0)
 generics           0.1.3       2022-07-05 [1] RSPM (R 4.2.2)
 GetoptLong         1.0.5       2020-12-15 [1] RSPM (R 4.2.2)
 ggalluvial         0.12.5      2023-07-30 [1] Github (corybrunson/ggalluvial@0b71a4d)
 ggbeeswarm         0.7.2       2023-04-29 [1] RSPM (R 4.3.0)
 ggmin              0.0.0.9000  2022-11-24 [1] Github (sjessa/ggmin@8ada274)
 ggnetwork          0.5.12      2023-04-28 [1] Github (briatte/ggnetwork@5b6ed24)
 ggplot2          * 3.4.2       2023-04-03 [1] RSPM (R 4.3.0)
 ggprism            1.0.4       2022-11-24 [1] Github (csdaw/ggprism@0e411f4)
 ggpubr             0.6.0.999   2023-04-28 [1] Github (kassambara/ggpubr@6aeb4f7)
 ggrastr            1.0.2       2023-06-01 [1] RSPM (R 4.3.0)
 ggrepel            0.9.3       2023-02-03 [1] RSPM (R 4.3.0)
 ggridges           0.5.4       2022-09-26 [1] RSPM (R 4.2.2)
 ggsignif           0.6.4.9000  2023-07-30 [1] Github (const-ae/ggsignif@080e119)
 ggstatsplot      * 0.11.1.9000 2023-04-28 [1] Github (IndrajeetPatil/ggstatsplot@befe812)
 GlobalOptions      0.1.2       2020-06-10 [1] RSPM (R 4.2.2)
 globals            0.16.2      2022-11-21 [1] RSPM (R 4.2.2)
 glue               1.6.2       2022-02-24 [1] RSPM (R 4.2.2)
 goftest            1.2-3       2021-10-07 [1] RSPM (R 4.2.2)
 gridBase           0.4-7       2014-02-24 [1] RSPM (R 4.3.0)
 gridExtra          2.3         2017-09-09 [1] RSPM (R 4.2.2)
 gtable             0.3.3       2023-03-21 [1] RSPM (R 4.3.0)
 hdf5r              1.3.8       2023-01-21 [1] RSPM (R 4.3.0)
 here             * 1.0.1       2020-12-13 [1] RSPM (R 4.2.2)
 hms                1.1.3       2023-03-21 [1] RSPM (R 4.3.0)
 htmltools          0.5.5       2023-03-23 [1] RSPM (R 4.3.0)
 htmlwidgets        1.6.2       2023-03-17 [1] RSPM (R 4.3.0)
 httpuv             1.6.11      2023-05-11 [1] RSPM (R 4.3.0)
 httr               1.4.6       2023-05-08 [1] RSPM (R 4.3.0)
 ica                1.0-3       2022-07-08 [1] RSPM (R 4.2.2)
 igraph           * 1.5.0.1     2023-07-23 [1] RSPM (R 4.3.1)
 insight            0.19.2      2023-05-23 [1] RSPM (R 4.3.0)
 IRanges            2.34.0      2023-04-25 [1] Bioconductor
 irlba              2.3.5.1     2022-10-03 [1] RSPM (R 4.2.2)
 iterators          1.0.14      2022-02-05 [1] RSPM (R 4.2.2)
 janitor            2.2.0.9000  2023-04-28 [1] Github (sfirke/janitor@d64c8bb)
 jsonlite           1.8.7       2023-06-29 [1] RSPM (R 4.3.1)
 KernSmooth         2.23-21     2023-05-03 [3] CRAN (R 4.3.1)
 knitr              1.43        2023-05-25 [1] RSPM (R 4.3.0)
 labeling           0.4.2       2020-10-20 [1] RSPM (R 4.2.2)
 later              1.3.1       2023-05-02 [1] RSPM (R 4.3.0)
 lattice            0.21-8      2023-04-05 [1] RSPM (R 4.3.0)
 lazyeval           0.2.2       2019-03-15 [1] RSPM (R 4.2.2)
 leiden             0.4.3       2022-09-10 [1] RSPM (R 4.2.2)
 lifecycle          1.0.3       2022-10-07 [1] RSPM (R 4.2.2)
 listenv            0.9.0       2022-12-16 [1] RSPM (R 4.3.0)
 lmtest             0.9-40      2022-03-21 [1] RSPM (R 4.2.2)
 lubridate        * 1.9.2       2023-02-10 [1] RSPM (R 4.3.0)
 magick             2.7.4       2023-03-09 [1] RSPM (R 4.3.0)
 magrittr         * 2.0.3       2022-03-30 [1] RSPM (R 4.2.2)
 MASS               7.3-60      2023-05-04 [1] RSPM (R 4.3.0)
 Matrix             1.6-0       2023-07-08 [1] RSPM (R 4.3.1)
 matrixStats        1.0.0       2023-06-02 [1] RSPM (R 4.3.0)
 mime               0.12        2021-09-28 [1] RSPM (R 4.2.2)
 miniUI             0.1.1.1     2018-05-18 [1] RSPM (R 4.2.2)
 multcomp           1.4-24      2023-06-14 [1] RSPM (R 4.3.0)
 munsell            0.5.0       2018-06-12 [1] RSPM (R 4.2.2)
 mvtnorm            1.2-2       2023-06-08 [1] RSPM (R 4.3.0)
 network            1.18.1      2023-01-24 [1] RSPM (R 4.3.0)
 nlme               3.1-162     2023-01-31 [3] CRAN (R 4.2.2)
 NMF                0.26        2023-03-20 [1] RSPM (R 4.3.0)
 paletteer          1.5.0       2022-10-19 [1] RSPM (R 4.2.2)
 parallelly         1.36.0      2023-05-26 [1] RSPM (R 4.3.0)
 parameters         0.21.1      2023-05-26 [1] RSPM (R 4.3.0)
 patchwork        * 1.1.2.9000  2023-04-28 [1] Github (thomasp85/patchwork@c14c960)
 pbapply            1.7-2       2023-06-27 [1] RSPM (R 4.3.1)
 pillar             1.9.0       2023-03-22 [1] RSPM (R 4.3.0)
 pkgconfig          2.0.3       2019-09-22 [1] RSPM (R 4.2.2)
 plotly             4.10.2      2023-06-03 [1] RSPM (R 4.3.0)
 plyr               1.8.8       2022-11-11 [1] CRAN (R 4.2.2)
 png                0.1-8       2022-11-29 [1] RSPM (R 4.3.0)
 polyclip           1.10-4      2022-10-20 [1] RSPM (R 4.2.2)
 progressr          0.13.0      2023-01-10 [1] RSPM (R 4.3.0)
 promises           1.2.0.1     2021-02-11 [1] RSPM (R 4.2.2)
 purrr            * 1.0.1       2023-01-10 [1] RSPM (R 4.3.0)
 R.methodsS3        1.8.2       2022-06-13 [1] RSPM (R 4.2.2)
 R.oo               1.25.0      2022-06-12 [1] RSPM (R 4.2.2)
 R.utils            2.12.2      2022-11-11 [1] CRAN (R 4.2.2)
 R6                 2.5.1       2021-08-19 [1] RSPM (R 4.2.2)
 ragg               1.2.5       2023-01-12 [1] RSPM (R 4.3.0)
 RANN               2.6.1       2019-01-08 [1] RSPM (R 4.2.2)
 rappdirs           0.3.3       2021-01-31 [1] RSPM (R 4.2.2)
 RColorBrewer     * 1.1-3       2022-04-03 [1] RSPM (R 4.2.2)
 Rcpp               1.0.11      2023-07-06 [1] RSPM (R 4.3.1)
 RcppAnnoy          0.0.20      2022-10-27 [1] RSPM (R 4.2.2)
 readr            * 2.1.4       2023-02-10 [1] RSPM (R 4.3.0)
 registry           0.5-1       2019-03-05 [1] RSPM (R 4.2.2)
 rematch2           2.1.2       2020-05-01 [1] RSPM (R 4.2.2)
 remotes            2.4.2       2021-11-30 [1] RSPM (R 4.2.2)
 repr               1.1.6       2023-01-26 [1] RSPM (R 4.3.0)
 reshape2           1.4.4       2020-04-09 [1] RSPM (R 4.2.2)
 reticulate       * 1.30        2023-06-09 [1] RSPM (R 4.3.0)
 rjson              0.2.21      2022-01-09 [1] RSPM (R 4.2.2)
 rlang              1.1.1       2023-04-28 [1] RSPM (R 4.3.0)
 rmarkdown          2.22        2023-06-01 [1] RSPM (R 4.3.0)
 rngtools           1.5.2       2021-09-20 [1] RSPM (R 4.3.0)
 ROCR               1.0-11      2020-05-02 [1] RSPM (R 4.2.2)
 rprojroot          2.0.3       2022-04-02 [1] RSPM (R 4.2.2)
 RSpectra           0.16-1      2022-04-24 [1] RSPM (R 4.2.2)
 rstatix            0.7.2       2023-02-01 [1] RSPM (R 4.3.0)
 rstudioapi         0.14        2022-08-22 [1] RSPM (R 4.2.2)
 rsvd               1.0.5       2021-04-16 [1] RSPM (R 4.2.2)
 Rtsne              0.16        2022-04-17 [1] RSPM (R 4.2.2)
 S4Vectors          0.38.1      2023-05-02 [1] Bioconductor
 sandwich           3.0-2       2022-06-15 [1] RSPM (R 4.3.0)
 scales             1.2.1       2022-08-20 [1] RSPM (R 4.2.2)
 scattermore        1.2         2023-06-12 [1] RSPM (R 4.3.0)
 scCustomize      * 1.1.1       2023-04-28 [1] Github (samuel-marsh/scCustomize@d08268d)
 sceasy           * 0.0.7       2022-11-24 [1] Github (cellgeni/sceasy@0cfc0e3)
 sctransform        0.3.5       2022-09-21 [1] RSPM (R 4.2.2)
 sessioninfo        1.2.2       2021-12-06 [1] RSPM (R 4.2.2)
 Seurat           * 4.3.0       2022-11-18 [1] RSPM (R 4.2.2)
 SeuratDisk       * 0.0.0.9020  2022-11-12 [1] Github (mojaveazure/seurat-disk@9b89970)
 SeuratObject     * 4.1.3       2022-11-07 [1] RSPM (R 4.2.2)
 SeuratWrappers   * 0.3.1       2022-11-12 [1] Github (satijalab/seurat-wrappers@d28512f)
 shape              1.4.6       2021-05-19 [1] RSPM (R 4.2.2)
 shiny              1.7.4       2022-12-15 [1] RSPM (R 4.3.0)
 skimr            * 2.1.5       2023-04-28 [1] Github (ropensci/skimr@d5126aa)
 sna                2.7-1       2023-01-24 [1] RSPM (R 4.3.0)
 snakecase          0.11.0      2019-05-25 [1] RSPM (R 4.2.2)
 sp                 1.6-1       2023-05-31 [1] RSPM (R 4.3.0)
 spatstat.data      3.0-1       2023-03-12 [1] RSPM (R 4.3.0)
 spatstat.explore   3.2-1       2023-05-13 [1] RSPM (R 4.3.0)
 spatstat.geom      3.2-1       2023-05-09 [1] RSPM (R 4.3.0)
 spatstat.random    3.1-5       2023-05-11 [1] RSPM (R 4.3.0)
 spatstat.sparse    3.0-1       2023-03-12 [1] RSPM (R 4.3.0)
 spatstat.utils     3.0-3       2023-05-09 [1] RSPM (R 4.3.0)
 statnet.common     4.9.0       2023-05-24 [1] RSPM (R 4.3.0)
 statsExpressions   1.5.1       2023-05-23 [1] RSPM (R 4.3.0)
 stringi            1.7.12      2023-01-11 [1] RSPM (R 4.3.0)
 stringr          * 1.5.0       2022-12-02 [1] RSPM (R 4.3.0)
 survival           3.5-5       2023-03-12 [1] RSPM (R 4.3.0)
 svglite            2.1.1       2023-01-10 [1] RSPM (R 4.3.0)
 systemfonts        1.0.4       2022-02-11 [1] RSPM (R 4.2.2)
 tensor             1.5         2012-05-05 [1] RSPM (R 4.2.2)
 textshaping        0.3.6       2021-10-13 [1] RSPM (R 4.2.2)
 TH.data            1.1-2       2023-04-17 [1] RSPM (R 4.3.0)
 tibble           * 3.2.1       2023-03-20 [1] RSPM (R 4.3.0)
 tidyr            * 1.3.0       2023-01-24 [1] RSPM (R 4.3.0)
 tidyselect         1.2.0       2022-10-10 [1] RSPM (R 4.2.2)
 tidyverse        * 2.0.0.9000  2023-04-28 [1] Github (tidyverse/tidyverse@8ec2e1f)
 timechange         0.2.0       2023-01-11 [1] RSPM (R 4.3.0)
 tzdb               0.4.0       2023-05-12 [1] RSPM (R 4.3.0)
 utf8               1.2.3       2023-01-31 [1] RSPM (R 4.3.0)
 uwot               0.1.14      2022-08-22 [1] RSPM (R 4.2.2)
 vctrs              0.6.3       2023-06-14 [1] RSPM (R 4.3.0)
 vipor              0.4.5       2017-03-22 [1] RSPM (R 4.2.2)
 viridis          * 0.6.3       2023-05-03 [1] RSPM (R 4.3.0)
 viridisLite      * 0.4.2       2023-05-02 [1] RSPM (R 4.3.0)
 vroom              1.6.3       2023-04-28 [1] RSPM (R 4.3.0)
 withr              2.5.0       2022-03-03 [1] RSPM (R 4.2.2)
 wordcloud          2.6         2018-08-24 [1] RSPM (R 4.3.0)
 xfun               0.39        2023-04-20 [1] RSPM (R 4.3.0)
 xtable             1.8-4       2019-04-21 [1] RSPM (R 4.2.2)
 yaml               2.3.7       2023-01-23 [1] RSPM (R 4.3.0)
 zeallot            0.1.0       2018-01-28 [1] RSPM (R 4.2.2)
 zoo                1.8-12      2023-04-13 [1] RSPM (R 4.3.0)

 [1] /usr/local/lib/R/site-library
 [2] /usr/lib/R/site-library
 [3] /usr/lib/R/library

─ Python configuration ───────────────────────────────────────────────────────
 python:         /home/etretiakov/micromamba/bin/python3
 libpython:      /home/etretiakov/micromamba/lib/libpython3.9.so
 pythonhome:     /home/etretiakov/micromamba:/home/etretiakov/micromamba
 version:        3.9.16 (main, Jan 11 2023, 16:05:54)  [GCC 11.2.0]
 numpy:          /home/etretiakov/micromamba/lib/python3.9/site-packages/numpy
 numpy_version:  1.23.5
 scanpy:         /home/etretiakov/micromamba/lib/python3.9/site-packages/scanpy
 
 python versions found: 
  /home/etretiakov/micromamba/bin/python3
  /home/etretiakov/micromamba/bin/python

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>