---
title: "Exploratory analysis and Quality Control of Arcuate Nucleus/Median Eminence dataset from Lutomska LM et al 2022"
author: "Evgenii O. Tretiakov"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  autodep        = TRUE,
  cache          = FALSE,
  cache.lazy     = FALSE,
  dev            = c("png", "pdf"),
  echo           = TRUE,
  error          = FALSE,
  fig.align      = "center",
  fig.width      = 14,
  fig.height     = 12,
  message        = FALSE,
  warning        = FALSE
)
Sys.setenv(RETICULATE_PYTHON = "/opt/python/3.8.8/bin/python")
# Load tidyverse infrastructure packages
suppressPackageStartupMessages({
  library(here)
  library(RColorBrewer)
  library(viridis)
  library(tidyverse)
  library(magrittr)
  library(stringr)
  library(skimr)
  library(future)
  library(zeallot)
  library(kableExtra)
  library(reticulate)
})
reticulate::use_condaenv("/opt/python/3.8.8/bin/python")

suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratWrappers)
  library(SeuratDisk)
  library(sctransform)
  library(glmGamPoi)
  library(patchwork)
  library(qs)
  library(Scillus)
  library(scCustomize)
  library(Nebulosa)
  library(mrtree)
  library(gprofiler2)
})

# Set paths
src_dir <- here("code")
data_dir <- here("data")
output_dir <- here("output")
plots_dir <- here(output_dir, "figures")
tables_dir <- here(output_dir, "tables")
source(here(src_dir, "genes.R"))
source(here(src_dir, "functions.R"))

# parallelisation
n_cores <- 16
plan("cluster", workers = n_cores)
options(future.globals.maxSize = 100000 * 1024^2) # 100Gb
# plan("sequential")
plan()

# set seed
reseed <- 42
set.seed(seed = reseed)

# ggplot2 theme
theme_set(ggmin::theme_powerpoint())
```

# Set QC parameters
```{r params}
low_cutoff_gene <- 500
high_cutoff_gene <- NULL
high_cutoff_gene <- 10000
low_cutoff_umis <- NULL
low_cutoff_umis <- -Inf
high_cutoff_umis <- 45000
high_cutoff_pc_mt <- 25
high_cutoff_pc_ribo <- 20
high_cutoff_pc_hb <- 0.1
high_cutoff_doublet_score <- NULL
high_cutoff_complexity <- 0.8
```

# Combined analysis of scRNA-seq datasets derived from the Arcuate nucleus

```{r load}
scrublet_SRR19578589 <-
  read_tsv(
    here("scrublet/SRR19578589", "SRR19578589_scrublet_calls_FPR_0.001.tsv")
  ) %>%
  mutate(
    origin = "SRR19578589",
    cell_name = str_c("SRR19578589_", barcode)
  )

scrublet_SRR19578590 <-
  read_tsv(
    here("scrublet/SRR19578590", "SRR19578590_scrublet_calls_FPR_0.001.tsv")
  ) %>%
  mutate(
    origin = "SRR19578590",
    cell_name = str_c("SRR19578590_", barcode)
  )

scrublet_SRR19578591 <-
  read_tsv(
    here("scrublet/SRR19578591", "SRR19578591_scrublet_calls_FPR_0.001.tsv")
  ) %>%
  mutate(
    origin = "SRR19578591",
    cell_name = str_c("SRR19578591_", barcode)
  )

scrublet <-
  bind_rows(
    scrublet_SRR19578589,
    scrublet_SRR19578590,
    scrublet_SRR19578591
  )

scrublet %>%
  janitor::tabyl(predicted_doublets, origin)


cell_bender_merged_0.001 <-
  Read_CellBender_h5_Multi_Directory(
    base_path = here("cellbender"),
    custom_name = "_output_FPR_0.001_filtered.h5",
    sample_names = sort(c("SRR19578589", "SRR19578590", "SRR19578591")), # must be sorted as the function internally doesn't index output of list.dir so can't reorder or subset
    merge = TRUE
  )

cell_bender_merged_0.01 <-
  Read_CellBender_h5_Multi_Directory(
    base_path = here("cellbender"),
    custom_name = "_output_FPR_0.01_filtered.h5",
    sample_names = sort(c("SRR19578589", "SRR19578590", "SRR19578591")), # must be sorted as the function internally doesn't index output of list.dir so can't reorder or subset
    merge = TRUE
  )
cell_bender_merged_0.05 <-
  Read_CellBender_h5_Multi_Directory(
    base_path = here("cellbender"),
    custom_name = "_output_FPR_0.05_filtered.h5",
    sample_names = sort(c("SRR19578589", "SRR19578590", "SRR19578591")), # must be sorted as the function internally doesn't index output of list.dir so can't reorder or subset
    merge = TRUE
  )

cell_bender_merged_0.1 <-
  Read_CellBender_h5_Multi_Directory(
    base_path = here("cellbender"),
    custom_name = "_output_FPR_0.1_filtered.h5",
    sample_names = sort(c("SRR19578589", "SRR19578590", "SRR19578591")), # must be sorted as the function internally doesn't index output of list.dir so can't reorder or subset
    merge = TRUE
  )

cell_ranger_merged <-
  Read10X_h5_Multi_Directory(
    base_path = here("cellranger"),
    default_10X_path = TRUE,
    h5_filename = "filtered_feature_bc_matrix.h5",
    merge = TRUE,
    sample_names = sort(c("SRR19578589", "SRR19578590", "SRR19578591")), # must be sorted as the function internally doesn't index output of list.dir so can't reorder or subset
    parallel = TRUE,
    num_cores = 16
  )

combined.srt_0.001 <-
  Create_CellBender_Merged_Seurat(
    raw_cell_bender_matrix = cell_bender_merged_0.001,
    raw_counts_matrix = cell_ranger_merged,
    raw_assay_name = "RAW"
  )

combined.srt_0.01 <-
  Create_CellBender_Merged_Seurat(
    raw_cell_bender_matrix = cell_bender_merged_0.01,
    raw_counts_matrix = cell_ranger_merged,
    raw_assay_name = "RAW"
  )

combined.srt_0.05 <-
  Create_CellBender_Merged_Seurat(
    raw_cell_bender_matrix = cell_bender_merged_0.05,
    raw_counts_matrix = cell_ranger_merged,
    raw_assay_name = "RAW"
  )

combined.srt_0.1 <-
  Create_CellBender_Merged_Seurat(
    raw_cell_bender_matrix = cell_bender_merged_0.1,
    raw_counts_matrix = cell_ranger_merged,
    raw_assay_name = "RAW"
  )
```


### Elimination of ambient RNA

#### FDR 0.001

```{r echo=FALSE}
combined.srt_0.001 <-
  Add_CellBender_Diff(
    seurat_object = combined.srt_0.001,
    raw_assay_name = "RAW",
    cell_bender_assay_name = "RNA"
  )

head(combined.srt_0.001@meta.data, 5) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```

```{r echo=FALSE}
median_stats_0.001 <-
  Median_Stats(
    seurat_object = combined.srt_0.001,
    group_by_var = "orig.ident",
    median_var = c("nCount_Diff", "nFeature_Diff")
  )

median_stats_0.001 %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(bootstrap_options = c("bordered", "condensed", "responsive", "striped"))
```

```{r echo=FALSE}
feature_diff_0.001 <-
  CellBender_Feature_Diff(
    seurat_object = combined.srt_0.001,
    raw_assay = "RAW",
    cell_bender_assay = "RNA"
  )

head(feature_diff_0.001, 20) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```


#### FDR 0.01

```{r echo=FALSE}
combined.srt_0.01 <-
  Add_CellBender_Diff(
    seurat_object = combined.srt_0.01,
    raw_assay_name = "RAW",
    cell_bender_assay_name = "RNA"
  )

head(combined.srt_0.01@meta.data, 5) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```

```{r echo=FALSE}
median_stats_0.01 <-
  Median_Stats(
    seurat_object = combined.srt_0.01,
    group_by_var = "orig.ident",
    median_var = c("nCount_Diff", "nFeature_Diff")
  )

median_stats_0.01 %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(bootstrap_options = c("bordered", "condensed", "responsive", "striped"))
```

```{r echo=FALSE}
feature_diff_0.01 <-
  CellBender_Feature_Diff(
    seurat_object = combined.srt_0.01,
    raw_assay = "RAW",
    cell_bender_assay = "RNA"
  )

head(feature_diff_0.01, 20) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```


#### FDR 0.05

```{r echo=FALSE}
combined.srt_0.05 <-
  Add_CellBender_Diff(
    seurat_object = combined.srt_0.05,
    raw_assay_name = "RAW",
    cell_bender_assay_name = "RNA"
  )

head(combined.srt_0.05@meta.data, 5) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```

```{r echo=FALSE}
median_stats_0.05 <-
  Median_Stats(
    seurat_object = combined.srt_0.05,
    group_by_var = "orig.ident",
    median_var = c("nCount_Diff", "nFeature_Diff")
  )

median_stats_0.05 %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(bootstrap_options = c("bordered", "condensed", "responsive", "striped"))
```

```{r echo=FALSE}
feature_diff_0.05 <-
  CellBender_Feature_Diff(
    seurat_object = combined.srt_0.05,
    raw_assay = "RAW",
    cell_bender_assay = "RNA"
  )

head(feature_diff_0.05, 20) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```


#### FDR 0.1

```{r echo=FALSE}
combined.srt_0.1 <-
  Add_CellBender_Diff(
    seurat_object = combined.srt_0.1,
    raw_assay_name = "RAW",
    cell_bender_assay_name = "RNA"
  )

head(combined.srt_0.1@meta.data, 5) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```

```{r echo=FALSE}
median_stats_0.1 <-
  Median_Stats(
    seurat_object = combined.srt_0.1,
    group_by_var = "orig.ident",
    median_var = c("nCount_Diff", "nFeature_Diff")
  )

median_stats_0.1 %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(bootstrap_options = c("bordered", "condensed", "responsive", "striped"))
```

```{r echo=FALSE}
feature_diff_0.1 <-
  CellBender_Feature_Diff(
    seurat_object = combined.srt_0.1,
    raw_assay = "RAW",
    cell_bender_assay = "RNA"
  )

head(feature_diff_0.1, 20) %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    )
  )
```



### Plot feature differences
In addition to returning the data.frame it can be useful to visually examine the changes/trends after running CellBender.

#### FDR 0.001

```{r echo=FALSE, fig.height=6, fig.width=8, fig.align='center'}
CellBender_Diff_Plot(feature_diff_df = feature_diff_0.001, num_labels = 50)
```


#### FDR 0.01

```{r echo=FALSE, fig.height=6, fig.width=8, fig.align='center'}
CellBender_Diff_Plot(feature_diff_df = feature_diff_0.01, num_labels = 50)
```


#### FDR 0.05

```{r echo=FALSE, fig.height=6, fig.width=8, fig.align='center'}
CellBender_Diff_Plot(feature_diff_df = feature_diff_0.05, num_labels = 50)
```


#### FDR 0.1

```{r echo=FALSE, fig.height=6, fig.width=8, fig.align='center'}
CellBender_Diff_Plot(feature_diff_df = feature_diff_0.1, num_labels = 50)
```

### Quality Control for different types of features
As we see it is optimal to select FDR 0.001 cutoff

```{r}
combined_srt <- combined.srt_0.001
median_stats <- median_stats_0.001
feature_diff <- feature_diff_0.001
rm(
  cell_bender_merged_0.001,
  combined.srt_0.001,
  median_stats_0.001,
  feature_diff_0.001,
  cell_bender_merged_0.01,
  combined.srt_0.01,
  median_stats_0.01,
  feature_diff_0.01,
  cell_bender_merged_0.05,
  combined.srt_0.05,
  median_stats_0.05,
  feature_diff_0.05,
  cell_bender_merged_0.1,
  combined.srt_0.1,
  median_stats_0.1,
  feature_diff_0.1
)
invisible(gc())
```



```{r pl-vln-qc, fig.align='center', fig.width=9, fig.asp = 0.618}
sex_genes <-
  str_to_title(c(
    "EHD2", "ESPL1", "JARID1D", "PNPLA4",
    "RPS4Y1", "XIST", "tsix", "Eif2s3y",
    "Ddx3y", "Uty", "Kdm5d"
  )) %>% .[. %in% rownames(combined_srt)]
stress_genes <-
  str_to_title(c(
    "Rpl26", "Gstp1", "Rpl35a", "Erh",
    "Slc25a5", "Pgk1", "Eno1",
    "Tubb2a", "Emc4", "Scg5"
  )) %>% .[. %in% rownames(combined_srt)]

combined_srt <-
  Store_Palette_Seurat(
    seurat_object = combined_srt,
    palette = rev(brewer.pal(n = 11, name = "RdYlGn")),
    palette_name = "mdat_Colour_Pal"
  )
combined_srt <-
  Store_Palette_Seurat(
    seurat_object = combined_srt,
    palette = rev(brewer.pal(n = 11, name = "Spectral")),
    palette_name = "expr_Colour_Pal"
  )

combined_srt <-
  Add_Mito_Ribo_Seurat(combined_srt, species = "mouse")
combined_srt[["percent_hb"]] <- PercentageFeatureSet(combined_srt, pattern = "^Hb[^(p)]")
combined_srt <-
  Add_Cell_Complexity_Seurat(combined_srt)

# Visualize QC metrics as a violin plot
p1 <-
  QC_Plots_Complexity(
    combined_srt,
    high_cutoff = high_cutoff_complexity,
    color_seed = reseed
  )
p2 <-
  QC_Plots_Genes(
    combined_srt,
    low_cutoff = low_cutoff_gene,
    high_cutoff = high_cutoff_gene,
    plot_title = "Genes per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p3 <-
  QC_Plots_UMIs(
    combined_srt,
    low_cutoff = low_cutoff_umis,
    high_cutoff = high_cutoff_umis,
    plot_title = "UMIs per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p4 <-
  QC_Plots_Mito(
    combined_srt,
    high_cutoff = high_cutoff_pc_mt,
    plot_title = "Mito genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p5 <-
  QC_Plots_Feature(
    combined_srt,
    feature = "percent_ribo",
    high_cutoff = high_cutoff_pc_ribo,
    y_axis_label = "% Ribosomal Genes Counts",
    plot_title = "Ribo genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p6 <-
  QC_Plots_Feature(
    combined_srt,
    feature = "percent_hb",
    high_cutoff = high_cutoff_pc_hb,
    y_axis_label = "% Hemoglobin Genes Counts",
    plot_title = "Hemoglobin genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )

wrap_plots(p1, p2, p3, p4, p5, p6, ncol = 3)
```


```{r pl-scatter-qc, fig.width=11, fig.asp = 0.8}
plot1 <- QC_Plot_GenevsFeature(seurat_object = combined_srt, feature1 = "percent_mito", low_cutoff_gene = low_cutoff_gene, high_cutoff_gene = high_cutoff_gene, high_cutoff_feature = high_cutoff_pc_mt, color_seed = reseed, ggplot_default_colors = T, pt.size = 4, shuffle_seed = reseed) & scale_y_log10()
plot2 <- QC_Plot_UMIvsGene(seurat_object = combined_srt, low_cutoff_gene = low_cutoff_gene, high_cutoff_gene = high_cutoff_gene, low_cutoff_UMI = low_cutoff_umis, high_cutoff_UMI = high_cutoff_umis, color_seed = reseed, ggplot_default_colors = T, pt.size = 4, shuffle_seed = reseed) & scale_x_log10() & scale_y_log10()
plot3 <- QC_Plot_GenevsFeature(seurat_object = combined_srt, feature1 = "percent_ribo", low_cutoff_gene = low_cutoff_gene, high_cutoff_gene = high_cutoff_gene, high_cutoff_feature = high_cutoff_pc_ribo, color_seed = reseed, ggplot_default_colors = T, pt.size = 4, shuffle_seed = reseed) & scale_y_log10()
plot4 <- FeatureScatter(combined_srt, feature1 = "percent_ribo", feature2 = "percent_mito", shuffle = T, pt.size = 4, seed = reseed)
(plot1 + plot2) / (plot3 + plot4)
```

```{r pl-scatter-qc-comb, fig.height=7, fig.width=13, fig.align='center'}
QC_Plot_UMIvsGene(
  seurat_object = combined_srt,
  meta_gradient_name = "percent_mito",
  low_cutoff_gene = low_cutoff_gene,
  high_cutoff_gene = high_cutoff_gene,
  high_cutoff_UMI = high_cutoff_umis,
  meta_gradient_low_cutoff = high_cutoff_pc_mt,
  meta_gradient_color = combined_srt@misc$mdat_Colour_Pal,
  combination = TRUE,
  color_seed = reseed,
  ggplot_default_colors = TRUE,
  pt.size = 4,
  shuffle_seed = reseed
) &
  scale_x_log10() & scale_y_log10()
```


```{r pl-scatter-doublets-log-prob, echo=FALSE, fig.height=7, fig.width=13, fig.align='center'}
combined_srt$cell_name <- colnames(combined_srt)
if (any(sum(combined_srt$cell_name %in% scrublet$cell_name))) {
  combined_srt@meta.data <-
    combined_srt@meta.data %>%
    left_join(scrublet,
      by = c("cell_name",
        "orig.ident" = "origin"
      )
    ) %>%
    mutate(QC = ifelse(
      test = predicted_doublets,
      yes = "Doublet",
      no = "Pass"
    ))
  rownames(combined_srt@meta.data) <- colnames(combined_srt)
  Split_FeatureScatter(
    seurat_object = combined_srt,
    feature1 = "nFeature_RNA",
    feature2 = "doublet_score",
    ggplot_default_colors = TRUE,
    color_seed = reseed,
    split.by = "orig.ident",
    group.by = "predicted_doublets",
    num_columns = 3,
    shuffle = TRUE,
    pt.size = 4,
    seed = reseed
  )
} else {
  combined_srt$QC <- "Pass"
}
```

```{r}
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$log10GenesPerUMI < high_cutoff_complexity &
      combined_srt@meta.data$QC == "Pass",
    "Low_Complexity",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$log10GenesPerUMI < high_cutoff_complexity &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "Low_Complexity",
    paste("Low_Complexity", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$nFeature_RNA < low_cutoff_gene &
      combined_srt@meta.data$QC == "Pass",
    "Low_nFeature",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$nFeature_RNA < low_cutoff_gene &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "Low_nFeature",
    paste("Low_nFeature", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_mito > high_cutoff_pc_mt &
      combined_srt@meta.data$QC == "Pass",
    "High_MT",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_mito > high_cutoff_pc_mt &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "High_MT",
    paste("High_MT", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$nCount_RNA > high_cutoff_umis &
      combined_srt@meta.data$QC == "Pass",
    "High_UMIs",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$nCount_RNA > high_cutoff_umis &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "High_UMIs",
    paste("High_UMIs", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_ribo > high_cutoff_pc_ribo &
      combined_srt@meta.data$QC == "Pass",
    "High_Ribo",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_ribo > high_cutoff_pc_ribo &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "High_Ribo",
    paste("High_Ribo", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_hb > high_cutoff_pc_hb &
      combined_srt@meta.data$QC == "Pass",
    "High_Hgb",
    combined_srt@meta.data$QC
  )
combined_srt$QC <-
  ifelse(
    combined_srt@meta.data$percent_hb > high_cutoff_pc_hb &
      combined_srt@meta.data$QC != "Pass" &
      combined_srt@meta.data$QC != "High_Hgb",
    paste("High_Hgb", combined_srt@meta.data$QC, sep = ","),
    combined_srt@meta.data$QC
  )
table(combined_srt$QC)
```

```{r pl-vln-qc-subset, fig.align='center', fig.width=9, fig.asp = 0.618}
# Visualize QC metrics as a violin plot again after subset
combined_subset_srt <- combined_srt
combined_subset_srt <- subset(combined_subset_srt, subset = QC == "Pass")
p1 <-
  QC_Plots_Complexity(
    seurat_object = combined_subset_srt,
    color_seed = reseed,
    ggplot_default_colors = T
  )
p2 <-
  QC_Plots_Genes(
    seurat_object = combined_subset_srt,
    low_cutoff = low_cutoff_gene,
    high_cutoff = high_cutoff_gene,
    plot_title = "Genes per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p3 <-
  QC_Plots_UMIs(
    seurat_object = combined_subset_srt,
    low_cutoff = low_cutoff_umis,
    high_cutoff = high_cutoff_umis,
    plot_title = "UMIs per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p4 <-
  QC_Plots_Mito(
    seurat_object = combined_subset_srt,
    high_cutoff = high_cutoff_pc_mt,
    plot_title = "Mito genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p5 <-
  QC_Plots_Feature(
    seurat_object = combined_subset_srt,
    feature = "percent_ribo",
    high_cutoff = high_cutoff_pc_ribo,
    y_axis_label = "% Ribosomal Genes Counts",
    plot_title = "Ribo genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p6 <-
  QC_Plots_Feature(
    seurat_object = combined_subset_srt,
    feature = "percent_hb",
    high_cutoff = high_cutoff_pc_hb,
    y_axis_label = "% Hemoglobin Genes Counts",
    plot_title = "Hemoglobin genes % per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )

wrap_plots(p1, p2, p3, p4, p5, p6, ncol = 3)
```

```{r pl-init-dr, fig.align='center', fig.asp=1.3, fig.width=9}
combined_srt <- NormalizeData(combined_srt)
combined_srt <-
  FindVariableFeatures(
    combined_srt,
    selection.method = "vst",
    nfeatures = 5000
  )
top100 <- head(VariableFeatures(combined_srt), 100)
plot5 <- VariableFeaturePlot(combined_srt)
LabelPoints(plot = plot5, points = top100, repel = TRUE, xnudge = 0, ynudge = 0)
all.genes <- rownames(combined_srt)
hvg <- VariableFeatures(combined_srt)
var_regex <- "^Hla-|^Ig[hjkl]|^Rna|^mt-|^Rp[sl]|^Hb[^(p)]|^Gm"
hvg <- hvg[str_detect(pattern = var_regex, string = hvg, negate = T)]
agg.genes <- GetAssayData(combined_srt, slot = "counts", assay = "RNA") |> rowSums()
all.genes <- all.genes[agg.genes > 0.001 * ncol(combined_srt)]
all.genes <- all.genes[str_detect(pattern = var_regex, string = all.genes, negate = T)]

combined_srt[["var_regex"]] <-
  PercentageFeatureSet(combined_srt, pattern = var_regex)

combined_srt <- ScaleData(combined_srt,
  features = all.genes,
  vars.to.regress = c(
    "log10GenesPerUMI"
  )
)

keep_genes <-
  c(gene_int, hvg) %>%
  unique() %>%
  .[!. %in% housekeeping_mouse] %>%
  .[!. %in% sex_genes] %>%
  .[!. %in% stress_genes]
glimpse(keep_genes)

out_of_hvg <- keep_genes[!keep_genes %in% hvg]
kable_material(
  kable(out_of_hvg, "html"),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)

hvg <- hvg[hvg %in% keep_genes]

plan("sequential")
invisible(gc())
plan("multicore", workers = 24)
n_cores <- 4
options(future.globals.maxSize = 32000 * 1024^2)

npcs <- 50
combined_srt <- RunPCA(combined_srt,
  features = hvg,
  npcs = npcs,
  seed.use = reseed,
  verbose = TRUE
)
VizDimLoadings(combined_srt, dims = 1:9, reduction = "pca") &
  theme(
    axis.text = element_text(size = 5),
    axis.title = element_text(size = 8, face = "bold")
  )
DimHeatmap(combined_srt, dims = 1:9, nfeatures = 20, cells = 500, balanced = T)
DimPlot_scCustom(combined_srt, reduction = "pca", color_seed = reseed, ggplot_default_colors = T, pt.size = 3)
ElbowPlot(combined_srt)

combined_srt <-
  JackStraw(
    object = combined_srt,
    assay = "RNA",
    reduction = "pca",
    dims = npcs,
    num.replicate = 100,
    prop.freq = 0.02,
    maxit = 1000
  )
combined_srt <-
  ScoreJackStraw(combined_srt,
    dims = seq_along(combined_srt[["pca"]]@stdev)
  )
JackStrawPlot(combined_srt, dims = seq_along(combined_srt[["pca"]]@stdev))
test_pc <-
  PCScore(
    object = combined_srt,
    PCs = seq_along(combined_srt[["pca"]]@stdev),
    score.thresh = 1e-05
  )
selected_pcs <-
  seq_along(combined_srt[["pca"]]@stdev)[
    test_pc$Score <= 1e-05 &
      (combined_srt[["pca"]]@stdev >
        quantile(combined_srt[["pca"]]@stdev, .10))
  ]
selected_pcs

combined_srt <-
  combined_srt |>
  FindNeighbors(
    dims = selected_pcs,
    k.param = 15,
    annoy.metric = "euclidean",
    n.trees = 100,
    verbose = FALSE
  ) |>
  RunUMAP(
    dims = selected_pcs,
    reduction.name = "umap",
    reduction.key = "UMAP_",
    return.model = FALSE,
    umap.method = "umap-learn",
    densmap = TRUE,
    dens.lambda = 1L,
    dens.frac = 0.3,
    n.epochs = 1000L,
    n.neighbors = 15L,
    min.dist = 0.01,
    spread = 2L,
    metric = "correlation",
    init = "pca",
    seed.use = reseed,
    verbose = FALSE
  )

metadata <- combined_srt@meta.data
rownames(metadata) <- colnames(combined_srt)
ref.labels <- metadata$orig.ident

resolutions <-
  modularity_event_sampling(
    A = combined_srt@graphs$RNA_snn,
    n.res = 30,
    gamma.min = 0.1,
    gamma.max = 3.000001
  ) # sample based on the similarity matrix

plan("sequential")
invisible(gc())
plan("multicore", workers = 24)
n_cores <- 4
options(future.globals.maxSize = 32000 * 1024^2)

# clustering using Suerat
combined_srt <- combined_srt |>
  FindClusters(
    algorithm = 4, method = "igraph",
    resolution = resolutions, random.seed = reseed,
    verbose = FALSE
  )

# initial cluster tree from Seurat flat clustering
plot_clustree(
  labelmat = combined_srt@meta.data,
  prefix = "RNA_snn_res.",
  ref.labels = ref.labels,
  plot.ref = FALSE
)
```

```{r update-gene-lists-rna}
source(here(src_dir, "genes.R"))
npr %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
np %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
gene_int %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
irs_genes %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
neurotrans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
glut %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
gaba %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
dopam %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
ach %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
mcr_genes %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
genes.embed %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
genes.nature %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
genes.jj %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
genes.anatomy.jj %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
astroenriched_mouse %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
astroprogenitor_humans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
astromature_humans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "scale.data"))]
```


```{r pl-mrtree-init, fig.align='center', fig.asp=.309, fig.width=9}
out <- mrtree(
  combined_srt,
  prefix = "RNA_snn_res.",
  n.cores = n_cores,
  consensus = FALSE,
  sample.weighted = TRUE,
  augment.path = FALSE,
  verbose = FALSE
)
# if there are few partitions per k, within resolution consensus step can speed up the algorithm
# weight per sample is encoraged if the classes are imbalanced

plot_tree(
  labelmat = out$labelmat.mrtree,
  ref.labels = ref.labels,
  plot.piechart = TRUE,
  node.size = 0.2,
  tip.label.dist = 10,
  bottom.margin = 30
)

# Adjusted Multiresolution Rand Index (AMRI)
ks.flat <- apply(
  out$labelmat.flat,
  2,
  FUN = function(x) {
    length(unique(x))
  }
)
ks.mrtree <- apply(
  out$labelmat.mrtree,
  2,
  FUN = function(x) {
    length(unique(x))
  }
)
amri.flat <- sapply(seq_len(ncol(out$labelmat.flat)), function(i) {
  AMRI(out$labelmat.flat[, i], ref.labels)$amri
})
amri.flat <- aggregate(amri.flat, by = list(k = ks.flat), FUN = mean)
amri.recon <- sapply(seq_len(ncol(out$labelmat.mrtree)), function(i) {
  AMRI(out$labelmat.mrtree[, i], ref.labels)$amri
})

df <- rbind(
  data.frame(
    k = amri.flat$k,
    amri = amri.flat$x,
    method = "Seurat flat"
  ),
  data.frame(k = ks.mrtree, amri = amri.recon, method = "MRtree")
)
ggplot2::ggplot(data = df, aes(x = k, y = amri, color = method)) +
  geom_line() +
  theme_bw()

stab.out <- stability_plot(out)
stab.out$plot

kable_material(
  kable(
    stab.out$df,
    "html"
  ),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)

resK <- SelectResolution(stab.out$df)
resK

kable_material(
  kable(
    table(
      out$labelmat.mrtree[, which.min(
        abs(as.integer(
          str_remove(dimnames(
            out$labelmat.mrtree
          )[[2]], "K")
        ) - resK)
      )]
    ),
    "html"
  ),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)

combined_srt$k_tree <- out$labelmat.mrtree[, which.min(
  abs(as.integer(
    str_remove(dimnames(
      out$labelmat.mrtree
    )[[2]], "K")
  ) - resK)
)] %>%
  as.numeric() %>%
  as.factor()
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
QC_Plots_Mito(
  combined_srt,
  high_cutoff = high_cutoff_pc_mt,
  plot_title = "Mito genes % per Nucleus (overclustered)",
  color_seed = reseed,
  ggplot_default_colors = T
)
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
QC_Plots_Feature(
  combined_srt,
  feature = "percent_ribo",
  high_cutoff = high_cutoff_pc_ribo,
  y_axis_label = "% Ribosomal Genes Counts",
  plot_title = "Ribo genes % per Nucleus (overclustered)",
  color_seed = reseed,
  ggplot_default_colors = T
)
```

```{r pl-umap-clusters, fig.align='center', fig.asp=.309, fig.width=9}
p1 <-
  DimPlot_scCustom(
    combined_srt,
    label = T, repel = T, pt.size = 2
  ) +
  ggtitle("Unsupervised overclustering") + NoLegend()
p2 <-
  DimPlot_scCustom(
    combined_srt,
    label = T, repel = T, group.by = "k_tree", pt.size = 2
  ) +
  ggtitle("MRTree") + NoLegend()

p1 | p2

Idents(combined_srt) <- "k_tree"
```


```{r fig.align='center', fig.asp=.618, fig.width=5}
FeaturePlot_scCustom(
  combined_srt,
  features = "percent_mito",
  colors_use = combined_srt@misc$mdat_Colour_Pal,
  na_cutoff = NA,
  pt.size = 4,
  order = TRUE,
  alpha_na_exp = 0.3,
  alpha_exp = 0.75
) &
  theme(plot.title = element_text(size = 16))

FeaturePlot_scCustom(
  combined_srt,
  features = "percent_mito",
  colors_use = combined_srt@misc$mdat_Colour_Pal,
  na_cutoff = 5,
  pt.size = 4,
  order = TRUE,
  alpha_na_exp = 0.3,
  alpha_exp = 0.75
) &
  theme(plot.title = element_text(size = 16))

FeaturePlot_scCustom(
  combined_srt,
  features = "nFeature_RNA",
  colors_use = combined_srt@misc$mdat_Colour_Pal,
  pt.size = 4,
  order = TRUE,
  alpha_na_exp = 0.3,
  alpha_exp = 0.75
) &
  theme(plot.title = element_text(size = 16))
```

### Dual Assay Plotting
For Cell Bender especially, but also potentially for other assays as well, it can be helpful during analysis to plot the corrected and uncorrected counts for given feature.

```{r include=FALSE, echo=FALSE}
combined_srt <- NormalizeData(combined_srt, assay = "RAW")
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Npy1r", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Npy2r", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Esr1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Prlr", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Otp", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Lxn", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Mfn2", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Gfap", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Nfia", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Cst3", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Slc38a1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Ntsr2", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Apoe", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Slit2", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Ndrg2", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Plcb1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Aldh1l1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Aldh1a1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Tafa1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Sgcd", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "S100b", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "S100a6", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Ntrk2", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Glul", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Slc1a3", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Aldoc", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r echo=FALSE, fig.height=4, fig.width=10, fig.align='center'}
FeaturePlot_DualAssay(seurat_object = combined_srt, features = "Gja1", assay1 = "RNA", assay2 = "RAW", colors_use = combined_srt@misc$expr_Colour_Pal, pt.size = 4, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 24))
```

```{r}
DimPlot_scCustom(combined_srt, pt.size = 3, group.by = "QC", repel = T, label = T, label.size = 5)
```

```{r}
DimPlot_scCustom(combined_srt, label.size = 5, repel = T, pt.size = 3, label = T)
```

```{r}
combined_srt <- subset(combined_srt, subset = QC == "Pass")
DimPlot_scCustom(combined_srt, label.size = 4, repel = T, pt.size = 3, label = T)
```


### Reevaluate after subsetting low-quality cells

```{r}
combined_srt$comb_clstr1 <- Idents(combined_srt)
s.genes <- gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
g2m.genes <- gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

combined_srt <-
  CellCycleScoring(combined_srt,
    s.features = s.genes,
    g2m.features = g2m.genes
  )
table(combined_srt[[]]$Phase)
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
FeaturePlot_scCustom(combined_srt, features = "percent_mito", label.size = 4, repel = T, pt.size = 3, label = T, colors_use = combined_srt@misc$mdat_Colour_Pal, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75) &
  theme(plot.title = element_text(size = 16))
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
QC_Plots_Mito(
  combined_srt,
  high_cutoff = high_cutoff_pc_mt,
  plot_title = "Mito genes % per Nucleus",
  color_seed = reseed,
  ggplot_default_colors = T
)
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
FeaturePlot_scCustom(combined_srt,
  features = "percent_ribo",
  label.size = 4, repel = T, pt.size = 3, label = T, colors_use = combined_srt@misc$mdat_Colour_Pal, order = TRUE, alpha_na_exp = 0.3, alpha_exp = 0.75
) &
  theme(plot.title = element_text(size = 16))
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
QC_Plots_Feature(
  combined_srt,
  feature = "percent_ribo",
  high_cutoff = high_cutoff_pc_ribo,
  y_axis_label = "% Ribosomal Genes Counts",
  plot_title = "Ribo genes % per Nucleus",
  color_seed = reseed,
  ggplot_default_colors = T
)
```

```{r fig.align='center', fig.asp=.618, fig.width=5}
p1 <-
  QC_Plots_Genes(
    combined_srt,
    low_cutoff = low_cutoff_gene,
    high_cutoff = high_cutoff_gene,
    plot_title = "Genes per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p2 <-
  QC_Plots_UMIs(
    combined_srt,
    low_cutoff = low_cutoff_umis,
    high_cutoff = high_cutoff_umis,
    plot_title = "UMIs per Nucleus",
    color_seed = reseed,
    ggplot_default_colors = T
  )
p1 | p2
```

```{r fig.align='center', fig.width=9, fig.asp = 0.309}
FeaturePlot_scCustom(
  combined_srt,
  features = c("S.Score", "G2M.Score"),
  label.size = 4,
  repel = T,
  pt.size = 3,
  label = T,
  colors_use = combined_srt@misc$mdat_Colour_Pal,
  na_cutoff = NA,
  order = TRUE,
  alpha_na_exp = 0.3,
  alpha_exp = 0.75
) &
  theme(plot.title = element_text(size = 16))
```

```{r fig.align='center', fig.width=7, fig.asp = 0.618}
VlnPlot(combined_srt,
  features = c("S.Score", "G2M.Score")
) &
  theme(plot.title = element_text(size = 16))
```

### Apply SCTransform pipeline

```{r normalisation}
# normalize and run dimensionality reduction on control dataset
npcs <- 30
metadata <- combined_srt@meta.data
rownames(metadata) <- colnames(combined_srt)
combined_srt <-
  SCTransform(
    combined_srt,
    vst.flavor = "v2",
    ncells = ncol(combined_srt),
    variable.features.n = 5500,
    vars.to.regress = c(
      "log10GenesPerUMI", "percent_mito",
      "S.Score", "G2M.Score"
    ),
    return.only.var.genes = FALSE,
    seed.use = reseed,
    verbose = FALSE
  )
hvg <- VariableFeatures(combined_srt)
var_regex <- "^Hla-|^Ig[hjkl]|^Rna|^mt-|^Rp[sl]|^Hb[^(p)]|^Gm"
hvg <- hvg[str_detect(pattern = var_regex, string = hvg, negate = T)]

keep_genes <-
  c(gene_int, hvg) %>%
  unique() %>%
  .[!. %in% housekeeping_mouse] %>%
  .[!. %in% sex_genes] %>%
  .[!. %in% stress_genes]
glimpse(keep_genes)

out_of_hvg <- keep_genes[!keep_genes %in% hvg]
kable_material(
  kable(out_of_hvg, "html"),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)

hvg <- hvg[hvg %in% keep_genes]

combined_srt <- combined_srt %>%
  RunPCA(features = hvg, npcs = npcs, seed.use = reseed, verbose = FALSE)
```

```{r update-gene-lists-sct}
source(here(src_dir, "genes.R"))
npr %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
np %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
irs_genes %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
neurotrans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
glut %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
gaba %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
dopam %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
ach %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
mcr_genes %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
genes.embed %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
genes.nature %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
genes.jj %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
genes.anatomy.jj %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
astroenriched_mouse %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
astroprogenitor_humans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
astromature_humans %<>% .[. %in% rownames(GetAssayData(combined_srt, slot = "data"))]
```


```{r pca-genes}
print(combined_srt[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r pl-pca-loadings, fig.asp=1.618, fig.width=8}
VizDimLoadings(combined_srt, dims = 1:4, reduction = "pca")
```

```{r pl-pca-heatmap, fig.height=8, fig.width=6}
DimHeatmap(combined_srt, dims = 1:15, cells = 500, balanced = TRUE)
```

```{r pl-elbow-pca, fig.height=4, fig.width=6}
ElbowPlot(combined_srt, ndims = npcs)
```

```{r pl-unsupervised}
combined_srt <-
  combined_srt |>
  FindNeighbors(
    dims = seq_along(combined_srt[["pca"]]@stdev),
    k.param = 20,
    annoy.metric = "euclidean",
    n.trees = 100,
    verbose = FALSE
  ) |>
  RunUMAP(
    dims = seq_along(combined_srt[["pca"]]@stdev),
    reduction.name = "umap",
    reduction.key = "UMAP_",
    return.model = TRUE,
    umap.method = "umap-learn",
    densmap = TRUE,
    dens.lambda = 1L,
    dens.frac = 0.1,
    n.epochs = 1000L,
    n.neighbors = 20L,
    min.dist = 0.01,
    spread = 4L,
    metric = "correlation",
    init = "pca",
    seed.use = reseed,
    verbose = FALSE
  )
```


Plot by source after clean up

```{r pl-all-RNA, fig.align='center', fig.width=9, fig.asp = 0.618}
plEmbCombBatch <- DimPlot_scCustom(combined_srt,
  reduction = "umap",
  group.by = "orig.ident", pt.size = 3,
  label = TRUE, repel = TRUE, seed = reseed,
  ggplot_default_colors = TRUE, color_seed = reseed,
  shuffle = TRUE
) + NoLegend()
plEmbCombBatch
```

```{r pl-clustree, fig.width=10, fig.asp=1.618}
metadata <- combined_srt@meta.data
rownames(metadata) <- colnames(combined_srt)
ref.labels <- metadata$k_tree

resolutions <-
  modularity_event_sampling(
    A = combined_srt@graphs$SCT_snn,
    n.res = 70,
    gamma.min = 0.05,
    gamma.max = 4.000001
  ) # sample based on the similarity matrix

# clustering using Suerat
combined_srt <- combined_srt |>
  FindClusters(
    algorithm = 4, method = "igraph",
    resolution = resolutions, random.seed = reseed,
    verbose = FALSE
  )

# initial cluster tree from Seurat flat clustering
plot_clustree(
  labelmat = combined_srt@meta.data,
  prefix = "SCT_snn_res.",
  ref.labels = ref.labels,
  plot.ref = FALSE
)
```

```{r pl-mrtree, fig.height=6, fig.width=11, echo=TRUE, include=FALSE}
out <- mrtree(
  combined_srt,
  prefix = "SCT_snn_res.",
  n.cores = n_cores,
  consensus = FALSE,
  sample.weighted = TRUE,
  augment.path = FALSE,
  verbose = FALSE
)
# if there are few partitions per k, within resolution consensus step can speed up the algorithm
# weight per sample is encoraged if the classes are imbalanced

plot_tree(
  labelmat = out$labelmat.mrtree,
  ref.labels = ref.labels,
  plot.piechart = TRUE,
  node.size = 0.2,
  tip.label.dist = 10,
  bottom.margin = 30
)
```

```{r pl-clustering-amri, fig.align='center', fig.width=4, fig.asp = 0.618}
# Adjusted Multiresolution Rand Index (AMRI)
ks.flat <- apply(
  out$labelmat.flat,
  2,
  FUN = function(x) {
    length(unique(x))
  }
)
ks.mrtree <- apply(
  out$labelmat.mrtree,
  2,
  FUN = function(x) {
    length(unique(x))
  }
)
amri.flat <- sapply(seq_len(ncol(out$labelmat.flat)), function(i) {
  AMRI(out$labelmat.flat[, i], ref.labels)$amri
})
amri.flat <- aggregate(amri.flat, by = list(k = ks.flat), FUN = mean)
amri.recon <- sapply(seq_len(ncol(out$labelmat.mrtree)), function(i) {
  AMRI(out$labelmat.mrtree[, i], ref.labels)$amri
})

df <- rbind(
  data.frame(
    k = amri.flat$k,
    amri = amri.flat$x,
    method = "Seurat flat"
  ),
  data.frame(k = ks.mrtree, amri = amri.recon, method = "MRtree")
)
ggplot2::ggplot(data = df, aes(x = k, y = amri, color = method)) +
  geom_line() +
  theme_bw()
```

```{r pl-clustering-resolution, fig.align='center', fig.width=4, fig.asp = 0.618}
stab.out <- stability_plot(out)
stab.out$plot
```

```{r select-resolution}
kable_material(
  kable(
    stab.out$df,
    "html"
  ),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)

resK <- SelectResolution(stab.out$df)
resK

kable_material(
  kable(
    table(
      out$labelmat.mrtree[, which.min(
        abs(as.integer(
          str_remove(dimnames(
            out$labelmat.mrtree
          )[[2]], "K")
        ) - resK)
      )]
    ),
    "html"
  ),
  bootstrap_options = c(
    "bordered",
    "condensed",
    "responsive",
    "striped"
  ),
  position = "left",
  font_size = 14
)
```

```{r pl-clustering, fig.align='center', fig.width=9, fig.asp = 0.309}
combined_srt$k_tree <- out$labelmat.mrtree[, which.min(
  abs(as.integer(
    str_remove(dimnames(
      out$labelmat.mrtree
    )[[2]], "K")
  ) - resK)
)] %>%
  as.numeric() %>%
  as.factor()
p1 <- DimPlot_scCustom(combined_srt, label = T, repel = T, pt.size = 2) + ggtitle("Unsupervised overclustering") + NoLegend()
p2 <- DimPlot_scCustom(combined_srt, label = T, repel = T, group.by = "k_tree", pt.size = 2) + ggtitle("MRTree") + NoLegend()

p1 | p2
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Npy1r", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Npy1r: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Npy2r", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Npy2r: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Esr1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Esr1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Prlr", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Prlr: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Otp", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Otp: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Lxn", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Lxn: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Mfn2", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Mfn2: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Gfap", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Gfap: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Nfia", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Nfia: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Cst3", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Cst3: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Slc38a1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Slc38a1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Ntsr2", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Ntsr2: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Apoe", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Apoe: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Slit2", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Slit2: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Ndrg2", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Ndrg2: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Plcb1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Plcb1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Aldh1l1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Aldh1l1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Aldh1a1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Aldh1a1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Tafa1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Tafa1: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Sgcd", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Sgcd: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "S100b", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("S100b: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "S100a6", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("S100a6: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Ntrk2", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Ntrk2: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Glul", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Glul: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Slc1a3", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Slc1a3: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Aldoc", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Aldoc: ") + theme(plot.title = element_text(size = 24))
```

```{r fig.align='center', fig.width=5, fig.asp = 0.618}
FeaturePlot_scCustom(combined_srt, "Gja1", pt.size = 2, order = T, colors_use = combined_srt@misc$expr_Colour_Pal, alpha_na_exp = 0.3, alpha_exp = 0.75) +
  ggtitle("Gja1: ") + theme(plot.title = element_text(size = 24))
```

```{r pl-jj-dotplot-sct, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "SCT",
  features = genes.jj,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "G")
)
```

```{r pl-jj-dotplot-rna, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "RNA",
  features = genes.jj,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E")
)
```

```{r pl-ajj-dotplot-sct, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "SCT",
  features = genes.anatomy.jj,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "G")
)
```

```{r pl-ajj-dotplot-rna, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "RNA",
  features = genes.anatomy.jj,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E")
)
```

```{r pl-nature-dotplot-sct, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "SCT",
  features = genes.nature,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "G")
)
```

```{r pl-nature-dotplot-rna, fig.asp=.618, fig.width=16}
DotPlot_scCustom(
  seurat_object = combined_srt,
  assay = "RNA",
  features = genes.nature,
  flip_axes = TRUE,
  x_lab_rotate = TRUE,
  colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E")
)
```

We see the spread of our targets across derived clusters, which isn't optimal. Lets see if we will see some significant hits with proper statistical testing.

```{r markers-tables}
Idents(combined_srt) <- "k_tree"
combined_srt <-
  PrepSCTFindMarkers(combined_srt, assay = "SCT")

markers.logreg <-
  FindAllMarkers(
    combined_srt,
    assay = "SCT",
    verbose = FALSE,
    random.seed = reseed,
    only.pos = TRUE,
    min.pct = 0.2,
    base = 10,
    logfc.threshold = 0.2,
    densify = TRUE,
    test.use = "LR"
  )
write_csv(
  markers.logreg,
  here(
    tables_dir,
    "lutomska2022_Arc-all-mrk_logreg-sct_combined.csv"
  )
)

markers.logreg %>%
  group_by(cluster) %>%
  slice_max(n = 20, order_by = avg_log10FC) %>%
  kable("html") %>%
  kable_material(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    ),
    position = "left",
    font_size = 14
  )


markers.wilcox <-
  FindAllMarkers(
    combined_srt,
    assay = "SCT",
    verbose = FALSE,
    random.seed = reseed,
    only.pos = TRUE,
    min.pct = 0.2,
    base = 10,
    logfc.threshold = 0.2,
    densify = TRUE,
    test.use = "wilcox"
  )
write_csv(
  markers.wilcox,
  here(
    tables_dir,
    "lutomska2022_Arc-all-mrk_wilcox-sct_combined.csv"
  )
)
markers.wilcox %>%
  group_by(cluster) %>%
  slice_max(n = 20, order_by = avg_log10FC) %>%
  kable("html") %>%
  kable_material(
    bootstrap_options = c(
      "bordered",
      "condensed",
      "responsive",
      "striped"
    ),
    position = "left",
    font_size = 14
  )
```

```{r include=FALSE, fig.align='center', fig.height=10, fig.width=10}
top5_markers <- Extract_Top_Markers(marker_dataframe = markers.logreg, num_genes = 5, named_vector = FALSE, make_unique = TRUE, rank_by = "avg_log10FC")
pl_clst_dotplot <- Clustered_DotPlot(seurat_object = combined_srt, features = top5_markers)

# markers
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = top5_markers, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_top5_umap", file_type = ".pdf")
# stress
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = stress_genes, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_stress_umap", file_type = ".pdf")
# sex
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = sex_genes, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
# jj
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = genes.jj, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
# anatomy jj
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = genes.anatomy.jj, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
# astro
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = genes.embed, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
# nature astro
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = genes.nature, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
# np receptors
Iterate_FeaturePlot_scCustom(seurat_object = combined_srt, gene_list = npr, single_pdf = T, colors_use = viridis(n = 30, alpha = .75, direction = -1, option = "E"), pt.size = 3, alpha_exp = 0.75, alpha_na_exp = 0.3, file_path = here(plots_dir), file_name = "/combined_sex_umap", file_type = ".pdf")
```

```{r pl_clst_dotplot, echo=FALSE, fig.align='center', fig.height=10, fig.width=10}
pl_clst_dotplot[[2]]
```

```{r pl-heatmap, fig.width = 18, fig.asp = 1.33}
markers.logreg %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log10FC) -> top10
DoHeatmap(combined_srt, features = top10$gene) + NoLegend()
```


```{r}
SaveH5Seurat(
  combined_srt,
  filename = here(
    data_dir,
    "PRJNA847050_clusters.h5Seurat"
  ),
  overwrite = TRUE
)

Convert(
  here(
    data_dir,
    "PRJNA847050_clusters.h5Seurat"
  ),
  dest = "h5ad",
  overwrite = TRUE
)
```
